<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Model construction functions · COBREXA.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script><link href="../../assets/favicon.ico" rel="icon" type="image/x-icon"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.svg" alt="COBREXA.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.svg" alt="COBREXA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">COBREXA.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../quickstart/">COBREXA.jl in 10 minutes</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../examples/">Contents</a></li><li><a class="tocitem" href="../../examples/01_loading/">Loading models</a></li><li><a class="tocitem" href="../../examples/02_convert_save/">Converting, modifying and saving models</a></li><li><a class="tocitem" href="../../examples/03_exploring/">Exploring model contents</a></li><li><a class="tocitem" href="../../examples/03b_accessors/">Generic accessors</a></li><li><a class="tocitem" href="../../examples/04_core_model/"><code>CoreModel</code> usage</a></li><li><a class="tocitem" href="../../examples/04_standardmodel/">Basic usage of <code>StandardModel</code></a></li><li><a class="tocitem" href="../../examples/04b_standardmodel_construction/">Model construction and modification</a></li><li><a class="tocitem" href="../../examples/05a_fba/">Flux balance analysis (FBA)</a></li><li><a class="tocitem" href="../../examples/05b_fba_mods/">Extending FBA with modifications</a></li><li><a class="tocitem" href="../../examples/06_fva/">Flux variability analysis (FVA)</a></li><li><a class="tocitem" href="../../examples/07_gene_deletion/">Gene knockouts</a></li><li><a class="tocitem" href="../../examples/07_restricting_reactions/">Restricting and disabling individual reactions</a></li><li><a class="tocitem" href="../../examples/08_pfba/">Parsimonious flux balance analysis (pFBA)</a></li><li><a class="tocitem" href="../../examples/09_loopless/">Loopless FBA</a></li><li><a class="tocitem" href="../../examples/10_crowding/">FBA with crowding</a></li><li><a class="tocitem" href="../../examples/11_growth/">Growth media analysis</a></li><li><a class="tocitem" href="../../examples/12_mmdf/">Maximum-minimum driving force analysis</a></li><li><a class="tocitem" href="../../examples/13_moma/">Minimization of metabolic adjustment (MOMA)</a></li><li><a class="tocitem" href="../../examples/14_smoment/">sMOMENT</a></li><li><a class="tocitem" href="../../examples/15_gecko/">GECKO</a></li><li><a class="tocitem" href="../../examples/16_hit_and_run/">Hit and run sampling</a></li><li><a class="tocitem" href="../../examples/17_envelopes/">Production envelopes</a></li></ul></li><li><span class="tocitem">Parallel, distributed and HPC processing</span><ul><li><a class="tocitem" href="../../distributed/">Contents</a></li><li><a class="tocitem" href="../../distributed/1_functions/">Parallel processing overview</a></li><li><a class="tocitem" href="../../distributed/2_parallel/">Local parallel processing</a></li><li><a class="tocitem" href="../../distributed/3_slurm/">Working in a HPC environment</a></li><li><a class="tocitem" href="../../distributed/4_serialized/">Faster model distribution</a></li></ul></li><li><span class="tocitem">Core concepts guide</span><ul><li><a class="tocitem" href="../../concepts/">Contents</a></li><li><a class="tocitem" href="../../concepts/1_screen/">Screening many model variants</a></li><li><a class="tocitem" href="../../concepts/2_modifications/">Writing custom optimizer modifications</a></li><li><a class="tocitem" href="../../concepts/3_custom_models/">Working with custom models</a></li><li><a class="tocitem" href="../../concepts/4_wrappers/">Extending the models</a></li></ul></li><li><span class="tocitem">Reference (Functions and types)</span><ul><li><a class="tocitem" href="../">Contents</a></li><li><a class="tocitem" href="../analysis/">Analysis functions</a></li><li><a class="tocitem" href="../base/">Base functions</a></li><li><a class="tocitem" href="../io/">Input and output</a></li><li class="is-active"><a class="tocitem" href>Model construction functions</a><ul class="internal"><li><a class="tocitem" href="#Functions-for-changing-the-models"><span>Functions for changing the models</span></a></li><li><a class="tocitem" href="#Variant-specifiers"><span>Variant specifiers</span></a></li></ul></li><li><a class="tocitem" href="../types/">Types</a></li><li><a class="tocitem" href="../utils/">Utilities</a></li></ul></li><li><a class="tocitem" href="../../howToContribute/">How to contribute</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Reference (Functions and types)</a></li><li class="is-active"><a href>Model construction functions</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Model construction functions</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/master/docs/src/functions/reconstruction.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Model-construction-functions"><a class="docs-heading-anchor" href="#Model-construction-functions">Model construction functions</a><a id="Model-construction-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Model-construction-functions" title="Permalink"></a></h1><h2 id="Functions-for-changing-the-models"><a class="docs-heading-anchor" href="#Functions-for-changing-the-models">Functions for changing the models</a><a id="Functions-for-changing-the-models-1"></a><a class="docs-heading-anchor-permalink" href="#Functions-for-changing-the-models" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reaction!-Tuple{CoreModel, Reaction}" href="#COBREXA.add_reaction!-Tuple{CoreModel, Reaction}"><code>COBREXA.add_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reaction!(model::CoreModel, rxn::Reaction)
</code></pre><p>Add <code>rxn</code> to <code>model</code>. The model must already contain the metabolites used by <code>rxn</code> in the model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions!-Tuple{CoreModel, Vector{Reaction}}" href="#COBREXA.add_reactions!-Tuple{CoreModel, Vector{Reaction}}"><code>COBREXA.add_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions!(model::CoreModel, rxns::Vector{Reaction})
</code></pre><p>Add <code>rxns</code> to <code>model</code> efficiently. The model must already contain the metabolites used by <code>rxns</code> in the model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{CoreModel, AbstractMatrix{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{String}, AbstractVector{String}}" href="#COBREXA.add_reactions-Tuple{CoreModel, AbstractMatrix{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{String}, AbstractVector{String}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m::CoreModel, Sp::AbstractMatrix{Float64}, b::AbstractVector{Float64}, c::AbstractVector{Float64}, xl::AbstractVector{Float64}, xu::AbstractVector{Float64}, rxns::AbstractVector{String}, mets::AbstractVector{String}; check_consistency) -&gt; Union{CoreModel, Tuple{CoreModel, Any, Any}}
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L140">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{CoreModel, AbstractMatrix{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}}" href="#COBREXA.add_reactions-Tuple{CoreModel, AbstractMatrix{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m::CoreModel, Sp::AbstractMatrix{Float64}, b::AbstractVector{Float64}, c::AbstractVector{Float64}, xl::AbstractVector{Float64}, xu::AbstractVector{Float64}; check_consistency) -&gt; Union{CoreModel, Tuple{CoreModel, Any, Any}}
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L94">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{CoreModel, AbstractVector{Float64}, AbstractVector{Float64}, AbstractFloat, AbstractFloat, AbstractFloat, String, AbstractVector{String}}" href="#COBREXA.add_reactions-Tuple{CoreModel, AbstractVector{Float64}, AbstractVector{Float64}, AbstractFloat, AbstractFloat, AbstractFloat, String, AbstractVector{String}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m::CoreModel, s::AbstractVector{Float64}, b::AbstractVector{Float64}, c::AbstractFloat, xl::AbstractFloat, xu::AbstractFloat, rxn::String, mets::AbstractVector{String}; check_consistency) -&gt; Union{CoreModel, Tuple{CoreModel, Vector{Int64}, Any}}
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L67">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{CoreModel, AbstractVector{Float64}, AbstractVector{Float64}, AbstractFloat, AbstractFloat, AbstractFloat}" href="#COBREXA.add_reactions-Tuple{CoreModel, AbstractVector{Float64}, AbstractVector{Float64}, AbstractFloat, AbstractFloat, AbstractFloat}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m::CoreModel, s::AbstractVector{Float64}, b::AbstractVector{Float64}, c::AbstractFloat, xl::AbstractFloat, xu::AbstractFloat; check_consistency) -&gt; Union{CoreModel, Tuple{CoreModel, Vector{Int64}, Any}}
</code></pre><p>Add reaction(s) to a <code>CoreModel</code> model <code>m</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{CoreModel, CoreModel}" href="#COBREXA.add_reactions-Tuple{CoreModel, CoreModel}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m1::CoreModel, m2::CoreModel; check_consistency) -&gt; Union{CoreModel, Tuple{CoreModel, Vector{Int64}, Vector{Int64}}}
</code></pre><p>Add all reactions from <code>m2</code> to <code>m1</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L121">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{CoreModel, Int64}" href="#COBREXA.change_bound!-Tuple{CoreModel, Int64}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::CoreModel,
    rxn_idx::Int;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, 123, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L258-L273">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{CoreModel, String}" href="#COBREXA.change_bound!-Tuple{CoreModel, String}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::CoreModel,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L282-L297">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{CoreModel, Int64}" href="#COBREXA.change_bound-Tuple{CoreModel, Int64}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::CoreModel,
    rxn_idx::Int;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, 123, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L270-L285">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{CoreModel, String}" href="#COBREXA.change_bound-Tuple{CoreModel, String}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::CoreModel,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L295-L310">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.change_bounds!-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::CoreModel,
    rxn_idxs::AbstractVector{Int64};
    lower = (nothing for _ = rxn_idxs),
    upper = (nothing for _ = rxn_idxs),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [123, 234], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L264-L279">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.change_bounds!-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::CoreModel,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L286-L301">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::CoreModel,
    rxn_idxs::AbstractVector{Int64};
    lower = (nothing for _ = rxn_idxs),
    upper = (nothing for _ = rxn_idxs),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [123, 234], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L274-L289">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::CoreModel,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L299-L314">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_objective!-Tuple{CoreModel, Int64}" href="#COBREXA.change_objective!-Tuple{CoreModel, Int64}"><code>COBREXA.change_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_objective!(model::CoreModel, rxn_idx::Int64)
</code></pre><p>Change objective function of a CoreModel to a single <code>1</code> at reaction index <code>rxn_idx</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L409">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_objective!-Tuple{CoreModel, String}" href="#COBREXA.change_objective!-Tuple{CoreModel, String}"><code>COBREXA.change_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_objective!(model::CoreModel, rxn_id::String)
</code></pre><p>Change objective function of a CoreModel to a single <code>1</code> at the given reaction ID.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L434">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_objective!-Tuple{CoreModel, Vector{Int64}}" href="#COBREXA.change_objective!-Tuple{CoreModel, Vector{Int64}}"><code>COBREXA.change_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_objective!(model::CoreModel, rxn_idxs::Vector{Int64}; weights)
</code></pre><p>Change the objective to reactions at given indexes, optionally specifying their <code>weights</code> in the same order. By default, all set weights are 1.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L393">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_objective!-Tuple{CoreModel, Vector{String}}" href="#COBREXA.change_objective!-Tuple{CoreModel, Vector{String}}"><code>COBREXA.change_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_objective!(model::CoreModel, rxn_ids::Vector{String}; weights)
</code></pre><p>Change objective of given reaction IDs, optionally specifying objective <code>weights</code> in the same order as <code>rxn_ids</code>. By default, all set weights are 1.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L417">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{CoreModel, Int64}" href="#COBREXA.remove_metabolite!-Tuple{CoreModel, Int64}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::CoreModel, metabolite_idx::Int)</code></pre><p>Remove metabolite from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L348-L353">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{CoreModel, String}" href="#COBREXA.remove_metabolite!-Tuple{CoreModel, String}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::CoreModel, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L377-L382">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{CoreModel, Int64}" href="#COBREXA.remove_metabolite-Tuple{CoreModel, Int64}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::CoreModel, metabolite_idx::Int)</code></pre><p>Remove metabolite from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L367-L372">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{CoreModel, String}" href="#COBREXA.remove_metabolite-Tuple{CoreModel, String}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::CoreModel, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L385-L390">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.remove_metabolites!-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::CoreModel, metabolite_idxs::AbstractVector{Int64})</code></pre><p>Remove metabolites from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L352-L357">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.remove_metabolites!-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::CoreModel, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L381-L386">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::CoreModel, metabolite_idxs::AbstractVector{Int64})</code></pre><p>Remove metabolites from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L371-L376">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::CoreModel, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L389-L394">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{CoreModel, Int64}" href="#COBREXA.remove_reaction!-Tuple{CoreModel, Int64}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::CoreModel, reaction_idx::Int)</code></pre><p>Remove reaction from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L308-L313">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{CoreModel, String}" href="#COBREXA.remove_reaction!-Tuple{CoreModel, String}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::CoreModel, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L332-L337">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{CoreModel, Int64}" href="#COBREXA.remove_reaction-Tuple{CoreModel, Int64}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::CoreModel, reaction_idx::Int)</code></pre><p>Remove reaction from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L322-L327">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{CoreModel, String}" href="#COBREXA.remove_reaction-Tuple{CoreModel, String}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::CoreModel, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L340-L345">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.remove_reactions!-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::CoreModel, reaction_idxs::AbstractVector{Int64})</code></pre><p>Remove reactions from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L312-L317">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.remove_reactions!-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::CoreModel, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L336-L341">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::CoreModel, reaction_idxs::AbstractVector{Int64})</code></pre><p>Remove reactions from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L326-L331">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::CoreModel, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L344-L349">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.verify_consistency-Union{Tuple{K}, Tuple{B}, Tuple{V}, Tuple{M}, Tuple{CoreModel, M, V, V, B, B, K, K, Any, Any}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, B&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}" href="#COBREXA.verify_consistency-Union{Tuple{K}, Tuple{B}, Tuple{V}, Tuple{M}, Tuple{CoreModel, M, V, V, B, B, K, K, Any, Any}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, B&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}"><code>COBREXA.verify_consistency</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">verify_consistency(m::CoreModel, Sp::AbstractMatrix{Float64}, b::AbstractVector{Float64}, c::AbstractVector{Float64}, xl::AbstractVector{Float64}, xu::AbstractVector{Float64}, names::AbstractVector{String}, mets::AbstractVector{String}, new_reactions, new_metabolites) -&gt; Tuple{Any, Any}
</code></pre><p>Check the consistency of given reactions with existing reactions in <code>m</code>.</p><p>TODO: work in progress, doesn&#39;t return consistency status.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModel.jl#L211">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_coupling_constraints!-Tuple{CoreCoupling, AbstractVector{Float64}, AbstractFloat, AbstractFloat}" href="#COBREXA.add_coupling_constraints!-Tuple{CoreCoupling, AbstractVector{Float64}, AbstractFloat, AbstractFloat}"><code>COBREXA.add_coupling_constraints!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_coupling_constraints!(m::CoreCoupling, c::AbstractVector{Float64}, cl::AbstractFloat, cu::AbstractFloat)
</code></pre><p>Overload for adding a single coupling constraint.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L150">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_coupling_constraints!-Union{Tuple{V}, Tuple{CoreCoupling, AbstractMatrix{Float64}, V, V}} where V&lt;:AbstractVector{Float64}" href="#COBREXA.add_coupling_constraints!-Union{Tuple{V}, Tuple{CoreCoupling, AbstractMatrix{Float64}, V, V}} where V&lt;:AbstractVector{Float64}"><code>COBREXA.add_coupling_constraints!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_coupling_constraints!(m::CoreCoupling, C::AbstractMatrix{Float64}, cl::AbstractVector{Float64}, cu::AbstractVector{Float64})
</code></pre><p>In-place add a single coupling constraint in form</p><pre><code class="language-none">    cₗ ≤ C x ≤ cᵤ</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L164">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_coupling_constraints-Tuple{CoreCoupling, Vararg{Any}}" href="#COBREXA.add_coupling_constraints-Tuple{CoreCoupling, Vararg{Any}}"><code>COBREXA.add_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_coupling_constraints(m::CoreCoupling, args...) -&gt; CoreCoupling
</code></pre><p>Add constraints of the following form to CoreCoupling and return the modified model.</p><p>The arguments are same as for in-place <a href="#COBREXA.add_coupling_constraints!-Tuple{CoreCoupling, AbstractVector{Float64}, AbstractFloat, AbstractFloat}"><code>add_coupling_constraints!</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L128">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_coupling_constraints-Tuple{CoreModel, Vararg{Any}}" href="#COBREXA.add_coupling_constraints-Tuple{CoreModel, Vararg{Any}}"><code>COBREXA.add_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_coupling_constraints(m::CoreModel, args...) -&gt; CoreModelCoupled
</code></pre><p>Add coupling constraints to a plain <a href="../types/#COBREXA.CoreModel"><code>CoreModel</code></a> (returns a <a href="../types/#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L142">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{CoreModelCoupled, CoreModel}" href="#COBREXA.add_reactions-Tuple{CoreModelCoupled, CoreModel}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m1::CoreModelCoupled, m2::CoreModel; check_consistency) -&gt; CoreModelCoupled
</code></pre><p>Add all reactions from <code>m2</code> to <code>m1</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L80">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{K}, Tuple{V2}, Tuple{V1}, Tuple{CoreModelCoupled, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat, String, K}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}" href="#COBREXA.add_reactions-Union{Tuple{K}, Tuple{V2}, Tuple{V1}, Tuple{CoreModelCoupled, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat, String, K}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m::CoreModelCoupled, s::AbstractVector{Float64}, b::AbstractVector{Float64}, c::AbstractFloat, xl::AbstractFloat, xu::AbstractFloat, rxn::String, mets::AbstractVector{String}; check_consistency) -&gt; CoreModelCoupled
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{K}, Tuple{V}, Tuple{M}, Tuple{CoreModelCoupled, M, V, V, V, V, K, K}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}" href="#COBREXA.add_reactions-Union{Tuple{K}, Tuple{V}, Tuple{M}, Tuple{CoreModelCoupled, M, V, V, V, V, K, K}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m::CoreModelCoupled, Sp::AbstractMatrix{Float64}, b::AbstractVector{Float64}, c::AbstractVector{Float64}, xl::AbstractVector{Float64}, xu::AbstractVector{Float64}, rxns::AbstractVector{String}, mets::AbstractVector{String}; check_consistency) -&gt; CoreModelCoupled
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L95">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{V2}, Tuple{V1}, Tuple{CoreModelCoupled, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}}" href="#COBREXA.add_reactions-Union{Tuple{V2}, Tuple{V1}, Tuple{CoreModelCoupled, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m::CoreModelCoupled, s::AbstractVector{Float64}, b::AbstractVector{Float64}, c::AbstractFloat, xl::AbstractFloat, xu::AbstractFloat; check_consistency) -&gt; CoreModelCoupled
</code></pre><p>Add reaction(s) to a <code>CoreModelCoupled</code> model <code>m</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{V}, Tuple{M}, Tuple{CoreModelCoupled, M, V, V, V, V}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}}" href="#COBREXA.add_reactions-Union{Tuple{V}, Tuple{M}, Tuple{CoreModelCoupled, M, V, V, V, V}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m::CoreModelCoupled, Sp::AbstractMatrix{Float64}, b::AbstractVector{Float64}, c::AbstractVector{Float64}, xl::AbstractVector{Float64}, xu::AbstractVector{Float64}; check_consistency) -&gt; CoreModelCoupled
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{CoreCoupling, Int64}" href="#COBREXA.change_bound!-Tuple{CoreCoupling, Int64}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::CoreCoupling,
    rxn_idx::Int;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, 123, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L258-L273">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{CoreCoupling, String}" href="#COBREXA.change_bound!-Tuple{CoreCoupling, String}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::CoreCoupling,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L266-L281">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{CoreCoupling, Int64}" href="#COBREXA.change_bound-Tuple{CoreCoupling, Int64}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::CoreCoupling,
    rxn_idx::Int;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, 123, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L274-L289">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{CoreCoupling, String}" href="#COBREXA.change_bound-Tuple{CoreCoupling, String}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::CoreCoupling,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L286-L301">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{CoreCoupling, AbstractVector{Int64}}" href="#COBREXA.change_bounds!-Tuple{CoreCoupling, AbstractVector{Int64}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::CoreCoupling,
    rxn_idxs::AbstractVector{Int64};
    lower = (nothing for _ = rxn_idxs),
    upper = (nothing for _ = rxn_idxs),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [123, 234], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L262-L277">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{CoreCoupling, AbstractVector{String}}" href="#COBREXA.change_bounds!-Tuple{CoreCoupling, AbstractVector{String}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::CoreCoupling,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L270-L285">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{CoreCoupling, AbstractVector{Int64}}" href="#COBREXA.change_bounds-Tuple{CoreCoupling, AbstractVector{Int64}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::CoreCoupling,
    rxn_idxs::AbstractVector{Int64};
    lower = (nothing for _ = rxn_idxs),
    upper = (nothing for _ = rxn_idxs),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [123, 234], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L280-L295">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{CoreCoupling, AbstractVector{String}}" href="#COBREXA.change_bounds-Tuple{CoreCoupling, AbstractVector{String}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::CoreCoupling,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L292-L307">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_coupling_bounds!-Union{Tuple{V}, Tuple{CoreCoupling, Vector{Int64}}} where V&lt;:AbstractVector{Float64}" href="#COBREXA.change_coupling_bounds!-Union{Tuple{V}, Tuple{CoreCoupling, Vector{Int64}}} where V&lt;:AbstractVector{Float64}"><code>COBREXA.change_coupling_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_coupling_bounds!(model::CoreCoupling, constraints::Vector{Int64}; cl, cu)
</code></pre><p>Change the lower and/or upper bounds (<code>cl</code> and <code>cu</code>) for the given list of coupling constraints.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L226">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_objective!-Tuple{CoreCoupling, Vararg{Any}}" href="#COBREXA.change_objective!-Tuple{CoreCoupling, Vararg{Any}}"><code>COBREXA.change_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_objective!(model::CoreCoupling, args...; kwargs...) -&gt; Any
</code></pre><p>Forwards arguments to <a href="#COBREXA.change_objective!-Tuple{CoreModel, Int64}"><code>change_objective!</code></a> of the internal model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L373">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_coupling_constraints!-Tuple{CoreCoupling, Int64}" href="#COBREXA.remove_coupling_constraints!-Tuple{CoreCoupling, Int64}"><code>COBREXA.remove_coupling_constraints!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_coupling_constraints!(m::CoreCoupling, constraint::Int64)
</code></pre><p>Removes a single coupling constraints from a <a href="../types/#COBREXA.CoreCoupling"><code>CoreCoupling</code></a> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L203">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_coupling_constraints!-Tuple{CoreCoupling, Vector{Int64}}" href="#COBREXA.remove_coupling_constraints!-Tuple{CoreCoupling, Vector{Int64}}"><code>COBREXA.remove_coupling_constraints!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_coupling_constraints!(m::CoreCoupling, constraints::Vector{Int64})
</code></pre><p>Removes a set of coupling constraints from a <a href="../types/#COBREXA.CoreCoupling"><code>CoreCoupling</code></a> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L212">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_coupling_constraints-Tuple{CoreCoupling, Vararg{Any}}" href="#COBREXA.remove_coupling_constraints-Tuple{CoreCoupling, Vararg{Any}}"><code>COBREXA.remove_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_coupling_constraints(m::CoreCoupling, args...) -&gt; CoreCoupling
</code></pre><p>Remove coupling constraints from the linear model, and return the modified model. Arguments are the same as for in-place version <a href="#COBREXA.remove_coupling_constraints!-Tuple{CoreCoupling, Int64}"><code>remove_coupling_constraints!</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L190">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{CoreCoupling, Int64}" href="#COBREXA.remove_metabolite!-Tuple{CoreCoupling, Int64}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::CoreCoupling, metabolite_idx::Int)</code></pre><p>Remove metabolite from the model of type <code>CoreCoupling</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L336-L341">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{CoreCoupling, String}" href="#COBREXA.remove_metabolite!-Tuple{CoreCoupling, String}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::CoreCoupling, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>CoreCoupling</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L357-L362">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{CoreCoupling, Int64}" href="#COBREXA.remove_metabolite-Tuple{CoreCoupling, Int64}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::CoreCoupling, metabolite_idx::Int)</code></pre><p>Remove metabolite from the model of type <code>CoreCoupling</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L347-L352">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{CoreCoupling, String}" href="#COBREXA.remove_metabolite-Tuple{CoreCoupling, String}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::CoreCoupling, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>CoreCoupling</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L365-L370">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{CoreCoupling, AbstractVector{Int64}}" href="#COBREXA.remove_metabolites!-Tuple{CoreCoupling, AbstractVector{Int64}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::CoreCoupling, metabolite_idxs::AbstractVector{Int64})</code></pre><p>Remove metabolites from the model of type <code>CoreCoupling</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L340-L345">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{CoreCoupling, AbstractVector{String}}" href="#COBREXA.remove_metabolites!-Tuple{CoreCoupling, AbstractVector{String}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::CoreCoupling, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>CoreCoupling</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L361-L366">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{CoreCoupling, AbstractVector{Int64}}" href="#COBREXA.remove_metabolites-Tuple{CoreCoupling, AbstractVector{Int64}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::CoreCoupling, metabolite_idxs::AbstractVector{Int64})</code></pre><p>Remove metabolites from the model of type <code>CoreCoupling</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L351-L356">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{CoreCoupling, AbstractVector{String}}" href="#COBREXA.remove_metabolites-Tuple{CoreCoupling, AbstractVector{String}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::CoreCoupling, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>CoreCoupling</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L369-L374">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{CoreCoupling, Int64}" href="#COBREXA.remove_reaction!-Tuple{CoreCoupling, Int64}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::CoreCoupling, reaction_idx::Int)</code></pre><p>Remove reaction from the model of type <code>CoreCoupling</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L298-L303">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{CoreCoupling, String}" href="#COBREXA.remove_reaction!-Tuple{CoreCoupling, String}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::CoreCoupling, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>CoreCoupling</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L320-L325">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{CoreCoupling, Int64}" href="#COBREXA.remove_reaction-Tuple{CoreCoupling, Int64}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::CoreCoupling, reaction_idx::Int)</code></pre><p>Remove reaction from the model of type <code>CoreCoupling</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L309-L314">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{CoreCoupling, String}" href="#COBREXA.remove_reaction-Tuple{CoreCoupling, String}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::CoreCoupling, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>CoreCoupling</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L328-L333">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{CoreCoupling, AbstractVector{Int64}}" href="#COBREXA.remove_reactions!-Tuple{CoreCoupling, AbstractVector{Int64}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::CoreCoupling, reaction_idxs::AbstractVector{Int64})</code></pre><p>Remove reactions from the model of type <code>CoreCoupling</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L302-L307">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{CoreCoupling, AbstractVector{String}}" href="#COBREXA.remove_reactions!-Tuple{CoreCoupling, AbstractVector{String}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::CoreCoupling, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>CoreCoupling</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L324-L329">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{CoreCoupling, AbstractVector{Int64}}" href="#COBREXA.remove_reactions-Tuple{CoreCoupling, AbstractVector{Int64}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::CoreCoupling, reaction_idxs::AbstractVector{Int64})</code></pre><p>Remove reactions from the model of type <code>CoreCoupling</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L313-L318">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{CoreCoupling, AbstractVector{String}}" href="#COBREXA.remove_reactions-Tuple{CoreCoupling, AbstractVector{String}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::CoreCoupling, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>CoreCoupling</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/CoreModelCoupled.jl#L332-L337">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._Stoichiometry" href="#COBREXA._Stoichiometry"><code>COBREXA._Stoichiometry</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct _Stoichiometry</code></pre><p>A small helper type for constructing reactions inline</p><p><strong>Fields</strong></p><ul><li><code>s::Dict{String, Float64}</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/Reaction.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.:+-Tuple{Union{Metabolite, COBREXA._Stoichiometry}, Union{Metabolite, COBREXA._Stoichiometry}}" href="#Base.:+-Tuple{Union{Metabolite, COBREXA._Stoichiometry}, Union{Metabolite, COBREXA._Stoichiometry}}"><code>Base.:+</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">+(a::Union{Metabolite, COBREXA._Stoichiometry}, b::Union{Metabolite, COBREXA._Stoichiometry}) -&gt; COBREXA._Stoichiometry
</code></pre><p>Shorthand for <code>metabolite1 + metabolite2</code>. Add 2 groups of <a href="../types/#COBREXA.Metabolite"><code>Metabolite</code></a>s together to form reactions inline. Use with <code>+</code>, <code>*</code>, <a href="#COBREXA.:→-Tuple{Union{Nothing, Metabolite, COBREXA._Stoichiometry}, Union{Nothing, Metabolite, COBREXA._Stoichiometry}}"><code>→</code></a> and similar operators.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/Reaction.jl#L20">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.:←-Tuple{Union{Nothing, Metabolite, COBREXA._Stoichiometry}, Union{Nothing, Metabolite, COBREXA._Stoichiometry}}" href="#COBREXA.:←-Tuple{Union{Nothing, Metabolite, COBREXA._Stoichiometry}, Union{Nothing, Metabolite, COBREXA._Stoichiometry}}"><code>COBREXA.:←</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">←(substrates::Union{Nothing, Metabolite, COBREXA._Stoichiometry}, products::Union{Nothing, Metabolite, COBREXA._Stoichiometry}) -&gt; Reaction
</code></pre><p>Shorthand for <code>substrates ← products</code>. Make a reverse-only <a href="../types/#COBREXA.Reaction"><code>Reaction</code></a> from <code>substrates</code> and <code>products</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/Reaction.jl#L58">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.:→-Tuple{Union{Nothing, Metabolite, COBREXA._Stoichiometry}, Union{Nothing, Metabolite, COBREXA._Stoichiometry}}" href="#COBREXA.:→-Tuple{Union{Nothing, Metabolite, COBREXA._Stoichiometry}, Union{Nothing, Metabolite, COBREXA._Stoichiometry}}"><code>COBREXA.:→</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">→(substrates::Union{Nothing, Metabolite, COBREXA._Stoichiometry}, products::Union{Nothing, Metabolite, COBREXA._Stoichiometry}) -&gt; Reaction
</code></pre><p>Shorthand for <code>substrates → products</code>. Make a forward-only <a href="../types/#COBREXA.Reaction"><code>Reaction</code></a> from <code>substrates</code> and <code>products</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/Reaction.jl#L49">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.:↔-Tuple{Union{Nothing, Metabolite, COBREXA._Stoichiometry}, Union{Nothing, Metabolite, COBREXA._Stoichiometry}}" href="#COBREXA.:↔-Tuple{Union{Nothing, Metabolite, COBREXA._Stoichiometry}, Union{Nothing, Metabolite, COBREXA._Stoichiometry}}"><code>COBREXA.:↔</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">↔(substrates::Union{Nothing, Metabolite, COBREXA._Stoichiometry}, products::Union{Nothing, Metabolite, COBREXA._Stoichiometry}) -&gt; Reaction
</code></pre><p>Shorthand for <code>substrates ↔ products</code>. Make a bidirectional (reversible) <a href="../types/#COBREXA.Reaction"><code>Reaction</code></a> from <code>substrates</code> and <code>products</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/Reaction.jl#L67">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._make_reaction_dict-Tuple{Any, Any}" href="#COBREXA._make_reaction_dict-Tuple{Any, Any}"><code>COBREXA._make_reaction_dict</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_make_reaction_dict(r, p) -&gt; Dict{String, Float64}
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/Reaction.jl#L38">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{Serialized, Vararg{Any}}" href="#COBREXA.add_reactions-Tuple{Serialized, Vararg{Any}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(model::Serialized, ...)</code></pre><p>Calls <a href="#COBREXA.add_reactions-Tuple{CoreModel, AbstractMatrix{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{String}, AbstractVector{String}}"><code>add_reactions</code></a> of the internal serialized model type. Returns the modified un-serialized model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/SerializedModel.jl#L4-L9">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{Serialized, Vararg{Any}}" href="#COBREXA.change_bound-Tuple{Serialized, Vararg{Any}}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(model::Serialized, ...)</code></pre><p>Calls <a href="#COBREXA.change_bound-Tuple{CoreModel, Int64}"><code>change_bound</code></a> of the internal serialized model type. Returns the modified un-serialized model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/SerializedModel.jl#L5-L10">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{Serialized, Vararg{Any}}" href="#COBREXA.change_bounds-Tuple{Serialized, Vararg{Any}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(model::Serialized, ...)</code></pre><p>Calls <a href="#COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{Int64}}"><code>change_bounds</code></a> of the internal serialized model type. Returns the modified un-serialized model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/SerializedModel.jl#L6-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{Serialized, Vararg{Any}}" href="#COBREXA.remove_metabolite-Tuple{Serialized, Vararg{Any}}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::Serialized, ...)</code></pre><p>Calls <a href="#COBREXA.remove_metabolite-Tuple{CoreModel, Int64}"><code>remove_metabolite</code></a> of the internal serialized model type. Returns the modified un-serialized model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/SerializedModel.jl#L7-L12">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{Serialized, Vararg{Any}}" href="#COBREXA.remove_metabolites-Tuple{Serialized, Vararg{Any}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::Serialized, ...)</code></pre><p>Calls <a href="#COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{Int64}}"><code>remove_metabolites</code></a> of the internal serialized model type. Returns the modified un-serialized model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/SerializedModel.jl#L8-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{Serialized, Vararg{Any}}" href="#COBREXA.remove_reaction-Tuple{Serialized, Vararg{Any}}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::Serialized, ...)</code></pre><p>Calls <a href="#COBREXA.remove_reaction-Tuple{CoreModel, Int64}"><code>remove_reaction</code></a> of the internal serialized model type. Returns the modified un-serialized model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/SerializedModel.jl#L9-L14">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{Serialized, Vararg{Any}}" href="#COBREXA.remove_reactions-Tuple{Serialized, Vararg{Any}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::Serialized, ...)</code></pre><p>Calls <a href="#COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{Int64}}"><code>remove_reactions</code></a> of the internal serialized model type. Returns the modified un-serialized model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/SerializedModel.jl#L10-L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.unwrap_serialized-Tuple{Serialized}" href="#COBREXA.unwrap_serialized-Tuple{Serialized}"><code>COBREXA.unwrap_serialized</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">unwrap_serialized(model::Serialized) -&gt; Union{Nothing, M} where M
</code></pre><p>Returns the model stored in the serialized structure.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/SerializedModel.jl#L12">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_gene!-Tuple{StandardModel, Gene}" href="#COBREXA.add_gene!-Tuple{StandardModel, Gene}"><code>COBREXA.add_gene!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_gene!(model::StandardModel, gene::Gene)
</code></pre><p>Add <code>gene</code> to <code>model</code> based on gene <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_genes!-Tuple{StandardModel, Vector{Gene}}" href="#COBREXA.add_genes!-Tuple{StandardModel, Vector{Gene}}"><code>COBREXA.add_genes!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_genes!(model::StandardModel, genes::Vector{Gene})
</code></pre><p>Add <code>genes</code> to <code>model</code> based on gene <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L39">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_metabolite!-Tuple{StandardModel, Metabolite}" href="#COBREXA.add_metabolite!-Tuple{StandardModel, Metabolite}"><code>COBREXA.add_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_metabolite!(model::StandardModel, met::Metabolite)
</code></pre><p>Add <code>met</code> to <code>model</code> based on metabolite <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_metabolites!-Tuple{StandardModel, Vector{Metabolite}}" href="#COBREXA.add_metabolites!-Tuple{StandardModel, Vector{Metabolite}}"><code>COBREXA.add_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_metabolites!(model::StandardModel, mets::Vector{Metabolite})
</code></pre><p>Add <code>mets</code> to <code>model</code> based on metabolite <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L20">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reaction!-Tuple{StandardModel, Reaction}" href="#COBREXA.add_reaction!-Tuple{StandardModel, Reaction}"><code>COBREXA.add_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reaction!(model::StandardModel, rxn::Reaction)
</code></pre><p>Add <code>rxn</code> to <code>model</code> based on reaction <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions!-Tuple{StandardModel, Vector{Reaction}}" href="#COBREXA.add_reactions!-Tuple{StandardModel, Vector{Reaction}}"><code>COBREXA.add_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions!(model::StandardModel, rxns::Vector{Reaction})
</code></pre><p>Add <code>rxns</code> to <code>model</code> based on reaction <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{StandardModel, String}" href="#COBREXA.change_bound!-Tuple{StandardModel, String}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::StandardModel,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L157-L172">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{StandardModel, String}" href="#COBREXA.change_bound-Tuple{StandardModel, String}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::StandardModel,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L169-L184">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.change_bounds!-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::StandardModel,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L163-L178">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.change_bounds-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::StandardModel,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L173-L188">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_objective!-Tuple{StandardModel, Vector{String}}" href="#COBREXA.change_objective!-Tuple{StandardModel, Vector{String}}"><code>COBREXA.change_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_objective!(model::StandardModel, rxn_ids::Vector{String}; weights)
</code></pre><p>Change the objective for <code>model</code> to reaction(s) with <code>rxn_ids</code>, optionally specifying their <code>weights</code>. By default, assume equal weights. If no objective exists in model, sets objective.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L238">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_gene!-Tuple{StandardModel, String}" href="#COBREXA.remove_gene!-Tuple{StandardModel, String}"><code>COBREXA.remove_gene!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_gene!(model::StandardModel, gid::String; knockout_reactions)
</code></pre><p>Remove gene with <code>id</code> from <code>model</code>. If <code>knockout_reactions</code> is true, then also constrain reactions that require the genes to function to carry zero flux.</p><p><strong>Example</strong></p><pre><code class="language-none">remove_gene!(model, &quot;g1&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L142">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_genes!-Tuple{StandardModel, Vector{String}}" href="#COBREXA.remove_genes!-Tuple{StandardModel, Vector{String}}"><code>COBREXA.remove_genes!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_genes!(model::StandardModel, gids::Vector{String}; knockout_reactions)
</code></pre><p>Remove all genes with <code>ids</code> from <code>model</code>. If <code>knockout_reactions</code> is true, then also constrain reactions that require the genes to function to carry zero flux.</p><p><strong>Example</strong></p><pre><code class="language-none">remove_genes!(model, [&quot;g1&quot;, &quot;g2&quot;])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L112">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{StandardModel, String}" href="#COBREXA.remove_metabolite!-Tuple{StandardModel, String}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::StandardModel, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>StandardModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L208-L213">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{StandardModel, String}" href="#COBREXA.remove_metabolite-Tuple{StandardModel, String}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::StandardModel, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>StandardModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L226-L231">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.remove_metabolites!-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::StandardModel, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>StandardModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L212-L217">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.remove_metabolites-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::StandardModel, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>StandardModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L230-L235">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{StandardModel, String}" href="#COBREXA.remove_reaction!-Tuple{StandardModel, String}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::StandardModel, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>StandardModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L183-L188">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{StandardModel, String}" href="#COBREXA.remove_reaction-Tuple{StandardModel, String}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::StandardModel, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>StandardModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L197-L202">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.remove_reactions!-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::StandardModel, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>StandardModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L192-L197">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.remove_reactions-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::StandardModel, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>StandardModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L201-L206">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.@add_reactions!-Tuple{Symbol, Expr}" href="#COBREXA.@add_reactions!-Tuple{Symbol, Expr}"><code>COBREXA.@add_reactions!</code></a> — <span class="docstring-category">Macro</span></header><section><div><p>Shortcut to add multiple reactions and their lower and upper bounds</p><p><strong>Call variants</strong></p><pre><code class="language-none">@add_reactions! model begin
    reaction_name, reaction
end

@add_reactions! model begin
    reaction_name, reaction, lower_bound
end

@add_reactions! model begin
    reaction_name, reaction, lower_bound, upper_bound
end</code></pre><p><strong>Examples</strong></p><pre><code class="language-none">@add_reactions! model begin
    &quot;v1&quot;, nothing → A, 0, 500
    &quot;v2&quot;, A ↔ B + C, -500
    &quot;v3&quot;, B + C → nothing
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/StandardModel.jl#L58">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_community_objective!-Tuple{CoreModel, Dict{String, Float64}}" href="#COBREXA.add_community_objective!-Tuple{CoreModel, Dict{String, Float64}}"><code>COBREXA.add_community_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_community_objective!(community::CoreModel, objective_mets_weights::Dict{String, Float64}; objective_id)
</code></pre><p>Add an objective column to the <code>community</code> model with optional id <code>objective_id</code>. Supply a dictionary mapping the string names of the objective metabolites to their weights in <code>objective_mets_weights</code>. Note, the weights are negated inside the function so that positive weights are seen as reagents/substrates, NOT products in the reaction equation.</p><p><strong>Example</strong></p><pre><code class="language-none">add_community_objective!(model, Dict(&quot;met1&quot;=&gt;1.0, &quot;met2&quot;=&gt;2.0))</code></pre><p>See also: <a href="#COBREXA.update_community_objective!-Tuple{CoreModel, String, Dict{String, Float64}}"><code>update_community_objective!</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_community_objective!-Tuple{StandardModel, Dict{String, Float64}}" href="#COBREXA.add_community_objective!-Tuple{StandardModel, Dict{String, Float64}}"><code>COBREXA.add_community_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_community_objective!(community::StandardModel, objective_mets_weights::Dict{String, Float64}; objective_id)
</code></pre><p>Variant of [<code>add_community_objective!</code>] that takes a <code>StandardModel</code> community model as input.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L39">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_model_with_exchanges!-Tuple{StandardModel, MetabolicModel, Dict{String, String}}" href="#COBREXA.add_model_with_exchanges!-Tuple{StandardModel, MetabolicModel, Dict{String, String}}"><code>COBREXA.add_model_with_exchanges!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_model_with_exchanges!(community::StandardModel, model::MetabolicModel, exchange_rxn_mets::Dict{String, String}; model_name, biomass_id)
</code></pre><p>The <code>StandardModel</code> variant of <a href="#COBREXA.add_model_with_exchanges-Tuple{CoreModel, MetabolicModel, Dict{String, String}}"><code>add_model_with_exchanges</code></a>, but is in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L469">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_model_with_exchanges-Tuple{CoreModel, MetabolicModel, Dict{String, String}}" href="#COBREXA.add_model_with_exchanges-Tuple{CoreModel, MetabolicModel, Dict{String, String}}"><code>COBREXA.add_model_with_exchanges</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_model_with_exchanges(community::CoreModel, model::MetabolicModel, exchange_rxn_mets::Dict{String, String}; model_name, biomass_id) -&gt; CoreModel
</code></pre><p>Add <code>model</code> to <code>community</code>, which is a pre-existing community model with exchange reactions and metabolites in the dictionary <code>exchange_rxn_mets</code>. The <code>model_name</code> is appended to each reaction and metabolite, see <a href="#COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Type{CoreModel}, Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel"><code>join_with_exchanges</code></a>. If <code>biomass_id</code> is specified then a biomass metabolite for <code>model</code> is also added to the resulting model. The column corresponding to the <code>biomass_id</code> reaction then produces this new biomass metabolite with unit coefficient. The exchange reactions and metabolites in <code>exchange_rxn_mets</code> must already exist in <code>community</code>. Always returns a new community model because it is more efficient than resizing all the matrices.</p><p>No in-place variant for <code>CoreModel</code>s exists yet.</p><p><strong>Example</strong></p><pre><code class="language-none">community = add_model_with_exchanges(community,
    model,
    exchange_rxn_mets;
    model_name=&quot;species_2&quot;,
    biomass_id=&quot;BIOMASS_Ecoli_core_w_GAM&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L359">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_model_with_exchanges-Tuple{StandardModel, MetabolicModel, Dict{String, String}}" href="#COBREXA.add_model_with_exchanges-Tuple{StandardModel, MetabolicModel, Dict{String, String}}"><code>COBREXA.add_model_with_exchanges</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_model_with_exchanges(community::StandardModel, model::MetabolicModel, exchange_rxn_mets::Dict{String, String}; model_name, biomass_id) -&gt; StandardModel
</code></pre><p>The <code>StandardModel</code> variant of <a href="#COBREXA.add_model_with_exchanges-Tuple{CoreModel, MetabolicModel, Dict{String, String}}"><code>add_model_with_exchanges</code></a>. Makes a deepcopy of <code>community</code> and calls the inplace variant of this function on that copy.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L520">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Type{CoreModel}, Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel" href="#COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Type{CoreModel}, Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel"><code>COBREXA.join_with_exchanges</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">join_with_exchanges(::Type{CoreModel}, models::Array{M&lt;:MetabolicModel, 1}, exchange_rxn_mets::Dict{String, String}; biomass_ids, model_names) -&gt; CoreModel
</code></pre><p>Return a <code>CoreModel</code> representing the community model of <code>models</code> joined through their exchange reactions and metabolites in the dictionary <code>exchange_rxn_mets</code>, which maps exchange reactions to their associated metabolite. These exchange reactions and metabolites link model metabolites to environmental metabolites and reactions. Optionally specify <code>model_names</code> to append a specific name to each reaction and metabolite of an organism for easier reference (default is <code>species_i</code> for each model index i in <code>models</code>). Note, the bounds of the environmental variables are all set to zero. Thus, to run a simulation you need to constrain them appropriately. All the other bounds are inherited from the models used to construct the community model.</p><p>If <code>biomass_ids</code> is supplied, then a community model is returned that has an extra reaction added to the end of the stoichiometric matrix (last column) that can be assigned as the objective reaction. It also creates biomass &quot;metabolites&quot; that can be used in this objective reaction. In the returned mode, these biomass metabolites are produced by the reaction corresponding to <code>biomass_ids</code> in each model respectively. Note, this reaction is unspecified, further action needs to be taken to specify it, e.g. assign weights to the last column of the stoichiometric matrix in the rows corresponding to the biomass metabolites.</p><p>To further clarify how this <code>join</code> works. Suppose you have 2 organisms with stoichiometric matrices S₁ and S₂ and you want to link them with <code>exchange_rxn_mets = Dict(er₁ =&gt; em₁, er₂ =&gt; em₂, er₃ =&gt; em₃, ...)</code>. Then a new community stoichiometric matrix is constructed that looks like this:</p><pre><code class="language-none">            _      er₁  er₂  er₃  ...  b_
           |S₁                           |
           |   S₂                        |
        em₁|                             |
S   =   em₂|                             |
        em₃|                             |
        ...|                             |
        bm₁|                             |
        bm₂|_                           _|
</code></pre><p>The exchange reactions in each model get linked to environmental metabolites, <code>emᵢ</code>, and these get linked to environmental exchanges, <code>erᵢ</code>. These <code>erᵢ</code> behave like normal single organism exchange reactions. When <code>biomass_ids</code> are supplied, each model&#39;s biomass reaction produces a pseudo-metabolite (<code>bmᵢ</code>). These can be weighted in column <code>b</code>, called the <code>community_biomass</code> reaction in the community model, if desired. Refer to the tutorial if this is unclear.</p><p><strong>Example</strong></p><pre><code class="language-none">m1 = load_model(core_model_path)
m2 = load_model(CoreModel, core_model_path)

# need to list ALL the exchanges that will form part of the entire model
exchange_rxn_mets = Dict(k =&gt; first(keys(reaction_stoichiometry(m1, ex_rxn)))
    for filter(looks_like_exchange_reaction, reactions(m1)))

biomass_ids = [&quot;BIOMASS_Ecoli_core_w_GAM&quot;, &quot;BIOMASS_Ecoli_core_w_GAM&quot;]

community = join_with_exchanges(
    CoreModel,
    [m1, m2],
    exchange_rxn_mets;
    biomass_ids = biomass_ids,
)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L116">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Type{StandardModel}, Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel" href="#COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Type{StandardModel}, Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel"><code>COBREXA.join_with_exchanges</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">join_with_exchanges(::Type{StandardModel}, models::Array{M&lt;:MetabolicModel, 1}, exchange_rxn_mets::Dict{String, String}; biomass_ids, model_names) -&gt; StandardModel
</code></pre><p>A variant of <a href="#COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Type{CoreModel}, Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel"><code>join_with_exchanges</code></a> that returns a <code>StandardModel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L304">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.update_community_objective!-Tuple{CoreModel, String, Dict{String, Float64}}" href="#COBREXA.update_community_objective!-Tuple{CoreModel, String, Dict{String, Float64}}"><code>COBREXA.update_community_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">update_community_objective!(community::CoreModel, objective_id::String, objective_mets_weights::Dict{String, Float64})
</code></pre><p>Update the weights for the objective column with id <code>objective_id</code> in <code>community</code> using <code>objective_mets_weights</code>, which maps metabolite ids to weights. The current weights are reset to 0 before being updated to the supplied weights. Note, the weights are negated inside the function so that the objective metabolites are seen as reagents/substrates, NOT products in the reaction equation.</p><p><strong>Example</strong></p><pre><code class="language-none">update_community_objective!(model, &quot;community_biomass&quot;, Dict(&quot;met1&quot;=&gt;1.0, &quot;met2&quot;=&gt;2.0))</code></pre><p>See also: <a href="#COBREXA.add_community_objective!-Tuple{CoreModel, Dict{String, Float64}}"><code>add_community_objective!</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L63">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.update_community_objective!-Tuple{StandardModel, String, Dict{String, Float64}}" href="#COBREXA.update_community_objective!-Tuple{StandardModel, String, Dict{String, Float64}}"><code>COBREXA.update_community_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">update_community_objective!(community::StandardModel, objective_id::String, objective_mets_weights::Dict{String, Float64})
</code></pre><p>Variant of [<code>update_community_objective!</code>] that takes a <code>StandardModel</code> community model as input.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/community.jl#L100">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_gecko-Tuple{}" href="#COBREXA.with_gecko-Tuple{}"><code>COBREXA.with_gecko</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_gecko(; kwargs...) -&gt; COBREXA.var&quot;#400#401&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Specifies a model variant which adds extra semantics of the Gecko algorithm, giving a <a href="../types/#COBREXA.GeckoModel"><code>GeckoModel</code></a>. The arguments are forwarded to <a href="../analysis/#COBREXA.make_gecko_model-Tuple{MetabolicModel}"><code>make_gecko_model</code></a>. Intended for usage with <a href="../analysis/#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/enzymes.jl#L10">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_smoment-Tuple{}" href="#COBREXA.with_smoment-Tuple{}"><code>COBREXA.with_smoment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_smoment(; kwargs...) -&gt; COBREXA.var&quot;#397#398&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Specifies a model variant which adds extra semantics of the sMOMENT algorithm, giving a <a href="../types/#COBREXA.SMomentModel"><code>SMomentModel</code></a>. The arguments are forwarded to <a href="../analysis/#COBREXA.make_smoment_model-Tuple{MetabolicModel}"><code>make_smoment_model</code></a>. Intended for usage with <a href="../analysis/#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/enzymes.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_expression_limits-Tuple{}" href="#COBREXA.with_expression_limits-Tuple{}"><code>COBREXA.with_expression_limits</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_expression_limits(; kwargs...) -&gt; COBREXA.var&quot;#403#404&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Specifies a model variant which adds extra semantics of the <a href="../types/#COBREXA.ExpressionLimitedModel"><code>ExpressionLimitedModel</code></a>, simulating the E-Flux algorithm. The arguments are forwarded to <a href="../analysis/#COBREXA.make_expression_limited_model-Tuple{MetabolicModel}"><code>make_expression_limited_model</code></a>. Intended for usage with <a href="../analysis/#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/expressions.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._universal_stoichiometry-Tuple{Vector{Reaction}, Vector{String}}" href="#COBREXA._universal_stoichiometry-Tuple{Vector{Reaction}, Vector{String}}"><code>COBREXA._universal_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_universal_stoichiometry(urxns::Vector{Reaction}, mids::Vector{String}) -&gt; NamedTuple{(:stoichiometry, :lbs, :ubs, :new_mids), Tuple{SparseArrays.SparseMatrixCSC{Float64, Int64}, Vector{Float64}, Vector{Float64}, Vector{String}}}
</code></pre><p>A helper function that constructs the stoichiometric matrix of a set of <code>universal_reactions</code>. The order of the metabolites is determined with <code>mids</code>, so that this stoichiometric matrix can be combined with another one.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/gapfill_minimum_reactions.jl#L169">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gapfill_minimum_reactions-Tuple{MetabolicModel, Vector{Reaction}, Any}" href="#COBREXA.gapfill_minimum_reactions-Tuple{MetabolicModel, Vector{Reaction}, Any}"><code>COBREXA.gapfill_minimum_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gapfill_minimum_reactions(model::MetabolicModel, universal_reactions::Vector{Reaction}, optimizer; objective_bounds, maximum_new_reactions, weights, modifications) -&gt; JuMP.Model
</code></pre><p>Find a minimal set of reactions from <code>universal_reactions</code> that should be added to <code>model</code> so that the model has a feasible solution with bounds on its objective function given in <code>objective_bounds</code>.  Weights of the added reactions may be specified in <code>weights</code> to prefer adding reactions with lower weights.</p><p>Internally, this builds and solves a mixed integer program, following the method of Reed et al. (Reed, Jennifer L., et al. &quot;Systems approach to refining genome annotation.&quot; <em>Proceedings of the National Academy of Sciences</em> (2006)).</p><p>The function returns a solved JuMP optimization model, with the boolean reaction inclusion indicators in variable vector <code>y</code>. Use <a href="#COBREXA.gapfilled_mask-Tuple{Any}"><code>gapfilled_mask</code></a> or <a href="#COBREXA.gapfilled_rids-Tuple{Any, Vector{Reaction}}"><code>gapfilled_rids</code></a> to collect the reaction information in Julia datatypes.</p><p>To reduce the uncertainty in the MILP solver (and likely reduce the complexity), you may put a limit on the size of the added reaction set in <code>maximum_new_reactions</code>.</p><p><strong>Common pitfalls</strong></p><p>If <a href="#COBREXA.gapfill_minimum_reactions-Tuple{MetabolicModel, Vector{Reaction}, Any}"><code>gapfill_minimum_reactions</code></a> is supposed to generate any reasonable output, the input model <em>MUST NOT</em> be feasible, otherwise there is &quot;no work to do&quot; and no reactions are added. Notably, an inactive model (the flux is zero) is considered to be feasible. If this is the case, <a href="#COBREXA.gapfilled_rids-Tuple{Any, Vector{Reaction}}"><code>gapfilled_rids</code></a> will return an empty vector (as opposed to <code>nothing</code>).</p><p>To prevent this, you may need to modify the model to disallow the trivial solutions (for example by putting a lower bound on reactions that you expect to be working in the solved model, in a similar manner like how the ATP maintenance reaction is bounded in E. Coli &quot;core&quot; model). The <code>objective_bounds</code> parameter makes this easier by directly placing a bound on the objective value of the model, which typically forces the model to be active.</p><p>The <code>maximum_new_reactions</code> parameter may have critical impact on performance in some solvers, because (in a general worst case) there is <code>2^maximum_new_reactions</code> model variants to be examined. Putting a hard limit on the reaction count may serve as a heuristic that helps the solver not to waste too much time solving impractically complex subproblems.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/gapfill_minimum_reactions.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gapfilled_mask-Tuple{Any}" href="#COBREXA.gapfilled_mask-Tuple{Any}"><code>COBREXA.gapfilled_mask</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gapfilled_mask(opt_model) -&gt; BitVector
</code></pre><p>Get a <code>BitVector</code> of added reactions from the model solved by <a href="#COBREXA.gapfill_minimum_reactions-Tuple{MetabolicModel, Vector{Reaction}, Any}"><code>gapfill_minimum_reactions</code></a>. The bit indexes correspond to the indexes of <code>universal_reactions</code> given to the gapfilling function. In case the model is not solved, this returns <code>nothing</code>.</p><p>If this function returns a zero vector (instead of <code>nothing</code>), it is very likely that the original model was already feasible and you may need to constraint it more. Refer to &quot;pitfalls&quot; section in the documentation of <a href="#COBREXA.gapfill_minimum_reactions-Tuple{MetabolicModel, Vector{Reaction}, Any}"><code>gapfill_minimum_reactions</code></a> for more details.</p><p><strong>Example</strong></p><pre><code class="language-none">gapfill_minimum_reactions(myModel, myReactions, Tulip.Optimizer) |&gt; gapfilled_mask</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/gapfill_minimum_reactions.jl#L120">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gapfilled_rids-Tuple{Any, Vector{Reaction}}" href="#COBREXA.gapfilled_rids-Tuple{Any, Vector{Reaction}}"><code>COBREXA.gapfilled_rids</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gapfilled_rids(opt_model, universal_reactions::Vector{Reaction}) -&gt; Vector{String}
</code></pre><p>Utility to extract a short vector of IDs of the reactions added by the gapfilling algorithm. Use with <code>opt_model</code> returned from <a href="#COBREXA.gapfill_minimum_reactions-Tuple{MetabolicModel, Vector{Reaction}, Any}"><code>gapfill_minimum_reactions</code></a>.</p><p>If this function returns an empty vector (instead of <code>nothing</code>), it is very likely that the original model was already feasible and you may need to constraint it more. Refer to &quot;pitfalls&quot; section in the documentation of <a href="#COBREXA.gapfill_minimum_reactions-Tuple{MetabolicModel, Vector{Reaction}, Any}"><code>gapfill_minimum_reactions</code></a> for more details.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/gapfill_minimum_reactions.jl#L140">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gapfilled_rids-Tuple{Vector{Reaction}}" href="#COBREXA.gapfilled_rids-Tuple{Vector{Reaction}}"><code>COBREXA.gapfilled_rids</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gapfilled_rids(universal_reactions::Vector{Reaction}) -&gt; COBREXA.var&quot;#414#415&quot;{Vector{Reaction}}
</code></pre><p>Overload of <a href="#COBREXA.gapfilled_rids-Tuple{Any, Vector{Reaction}}"><code>gapfilled_rids</code></a> that can be piped easily.</p><p><strong>Example</strong></p><pre><code class="language-none">gapfill_minimum_reactions(myModel, myReactions, Tulip.Optimizer) |&gt; gapfilled_rids(myReactions)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/gapfill_minimum_reactions.jl#L157">source</a></section></article><h2 id="Variant-specifiers"><a class="docs-heading-anchor" href="#Variant-specifiers">Variant specifiers</a><a id="Variant-specifiers-1"></a><a class="docs-heading-anchor-permalink" href="#Variant-specifiers" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_added_reactions-Tuple" href="#COBREXA.with_added_reactions-Tuple"><code>COBREXA.with_added_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_added_reactions(args...; kwargs...) -&gt; COBREXA.var&quot;#448#449&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Specifies a model variant with reactions added. Forwards the arguments to <a href="#COBREXA.add_reactions-Tuple{CoreModel, AbstractMatrix{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{Float64}, AbstractVector{String}, AbstractVector{String}}"><code>add_reactions</code></a>. Intended to be used with <a href="../analysis/#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/modifications/generic.jl#L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_changed_bound-Tuple" href="#COBREXA.with_changed_bound-Tuple"><code>COBREXA.with_changed_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_changed_bound(args...; kwargs...) -&gt; COBREXA.var&quot;#436#437&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Specifies a model variant that has a new bound set. Forwards arguments to <a href="#COBREXA.change_bound-Tuple{CoreModel, Int64}"><code>change_bound</code></a>. Intended for usage with <a href="../analysis/#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/modifications/generic.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_changed_bounds-Tuple" href="#COBREXA.with_changed_bounds-Tuple"><code>COBREXA.with_changed_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_changed_bounds(args...; kwargs...) -&gt; COBREXA.var&quot;#439#440&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Specifies a model variant that has new bounds set. Forwards arguments to <a href="#COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{Int64}}"><code>change_bounds</code></a>. Intended for usage with <a href="../analysis/#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/modifications/generic.jl#L9">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_removed_metabolite-Tuple" href="#COBREXA.with_removed_metabolite-Tuple"><code>COBREXA.with_removed_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_removed_metabolite(args...; kwargs...) -&gt; COBREXA.var&quot;#442#443&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Specifies a model variant without a certain metabolite. Forwards arguments to <a href="#COBREXA.remove_metabolite-Tuple{CoreModel, Int64}"><code>remove_metabolite</code></a>. Intended to be used with <a href="../analysis/#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/modifications/generic.jl#L17">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_removed_metabolites-Tuple" href="#COBREXA.with_removed_metabolites-Tuple"><code>COBREXA.with_removed_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_removed_metabolites(args...; kwargs...) -&gt; COBREXA.var&quot;#445#446&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Plural version of <a href="#COBREXA.with_removed_metabolite-Tuple"><code>with_removed_metabolite</code></a>, calls <a href="#COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{Int64}}"><code>remove_metabolites</code></a> internally.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/modifications/generic.jl#L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_removed_reaction-Tuple" href="#COBREXA.with_removed_reaction-Tuple"><code>COBREXA.with_removed_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_removed_reaction(args...; kwargs...) -&gt; COBREXA.var&quot;#451#452&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Specifies a model variant without a certain reaction. Forwards arguments to <a href="#COBREXA.remove_reaction-Tuple{CoreModel, Int64}"><code>remove_reaction</code></a>. Intended to be used with <a href="../analysis/#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/modifications/generic.jl#L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_removed_reactions-Tuple" href="#COBREXA.with_removed_reactions-Tuple"><code>COBREXA.with_removed_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_removed_reactions(args...; kwargs...) -&gt; COBREXA.var&quot;#454#455&quot;{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}
</code></pre><p>Plural version of <a href="#COBREXA.with_removed_reaction-Tuple"><code>with_removed_reaction</code></a>, calls <a href="#COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{Int64}}"><code>remove_reactions</code></a> internally.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/28451319de55e210aa8f87ea89eacafaef79aa95/src/reconstruction/modifications/generic.jl#L50">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../io/">« Input and output</a><a class="docs-footer-nextpage" href="../types/">Types »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 31 October 2023 12:25">Tuesday 31 October 2023</span>. Using Julia version 1.8.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
