<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Types · COBREXA.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script><link href="../../assets/favicon.ico" rel="icon" type="image/x-icon"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.svg" alt="COBREXA.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.svg" alt="COBREXA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">COBREXA.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../quickstart/">COBREXA.jl in 10 minutes</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../examples/">Contents</a></li><li><a class="tocitem" href="../../examples/01_loading/">Loading models</a></li><li><a class="tocitem" href="../../examples/02_convert_save/">Converting, modifying and saving models</a></li><li><a class="tocitem" href="../../examples/03_exploring/">Exploring model contents</a></li><li><a class="tocitem" href="../../examples/03b_accessors/">Generic accessors</a></li><li><a class="tocitem" href="../../examples/04_core_model/"><code>CoreModel</code> usage</a></li><li><a class="tocitem" href="../../examples/04_standardmodel/">Basic usage of <code>StandardModel</code></a></li><li><a class="tocitem" href="../../examples/04b_standardmodel_construction/">Model construction and modification</a></li><li><a class="tocitem" href="../../examples/05a_fba/">Flux balance analysis (FBA)</a></li><li><a class="tocitem" href="../../examples/05b_fba_mods/">Extending FBA with modifications</a></li><li><a class="tocitem" href="../../examples/06_fva/">Flux variability analysis (FVA)</a></li><li><a class="tocitem" href="../../examples/07_gene_deletion/">Gene knockouts</a></li><li><a class="tocitem" href="../../examples/07_restricting_reactions/">Restricting and disabling individual reactions</a></li><li><a class="tocitem" href="../../examples/08_pfba/">Parsimonious flux balance analysis (pFBA)</a></li><li><a class="tocitem" href="../../examples/09_loopless/">Loopless FBA</a></li><li><a class="tocitem" href="../../examples/10_crowding/">FBA with crowding</a></li><li><a class="tocitem" href="../../examples/11_growth/">Growth media analysis</a></li><li><a class="tocitem" href="../../examples/12_mmdf/">Maximum-minimum driving force analysis</a></li><li><a class="tocitem" href="../../examples/13_moma/">Minimization of metabolic adjustment (MOMA)</a></li><li><a class="tocitem" href="../../examples/14_smoment/">sMOMENT</a></li><li><a class="tocitem" href="../../examples/15_gecko/">GECKO</a></li><li><a class="tocitem" href="../../examples/16_hit_and_run/">Hit and run sampling</a></li><li><a class="tocitem" href="../../examples/17_envelopes/">Production envelopes</a></li></ul></li><li><span class="tocitem">Parallel, distributed and HPC processing</span><ul><li><a class="tocitem" href="../../distributed/">Contents</a></li><li><a class="tocitem" href="../../distributed/1_functions/">Parallel processing overview</a></li><li><a class="tocitem" href="../../distributed/2_parallel/">Local parallel processing</a></li><li><a class="tocitem" href="../../distributed/3_slurm/">Working in a HPC environment</a></li><li><a class="tocitem" href="../../distributed/4_serialized/">Faster model distribution</a></li></ul></li><li><span class="tocitem">Core concepts guide</span><ul><li><a class="tocitem" href="../../concepts/">Contents</a></li><li><a class="tocitem" href="../../concepts/1_screen/">Screening many model variants</a></li><li><a class="tocitem" href="../../concepts/2_modifications/">Writing custom optimizer modifications</a></li><li><a class="tocitem" href="../../concepts/3_custom_models/">Working with custom models</a></li><li><a class="tocitem" href="../../concepts/4_wrappers/">Extending the models</a></li></ul></li><li><span class="tocitem">Reference (Functions and types)</span><ul><li><a class="tocitem" href="../">Contents</a></li><li><a class="tocitem" href="../analysis/">Analysis functions</a></li><li><a class="tocitem" href="../base/">Base functions</a></li><li><a class="tocitem" href="../io/">Input and output</a></li><li><a class="tocitem" href="../reconstruction/">Model construction functions</a></li><li class="is-active"><a class="tocitem" href>Types</a><ul class="internal"><li><a class="tocitem" href="#Base-types"><span>Base types</span></a></li><li><a class="tocitem" href="#Model-types-and-contents"><span>Model types and contents</span></a></li><li><a class="tocitem" href="#Model-type-wrappers"><span>Model type wrappers</span></a></li></ul></li><li><a class="tocitem" href="../utils/">Utilities</a></li></ul></li><li><a class="tocitem" href="../../howToContribute/">How to contribute</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Reference (Functions and types)</a></li><li class="is-active"><a href>Types</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Types</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/master/docs/src/functions/types.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h1><h2 id="Base-types"><a class="docs-heading-anchor" href="#Base-types">Base types</a><a id="Base-types-1"></a><a class="docs-heading-anchor-permalink" href="#Base-types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Maybe" href="#COBREXA.Maybe"><code>COBREXA.Maybe</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Maybe{T} = Union{Nothing, T}</code></pre><p>A nice name for &quot;nullable&quot; type.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/Maybe.jl#L2-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._default-Union{Tuple{T}, Tuple{T, Union{Nothing, T}}} where T" href="#COBREXA._default-Union{Tuple{T}, Tuple{T, Union{Nothing, T}}} where T"><code>COBREXA._default</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_default(d, x::Union{Nothing, T} where T) -&gt; Any
</code></pre><p>Fold the <code>Maybe{T}</code> down to <code>T</code> by defaulting.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/Maybe.jl#L9">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._maybemap-Tuple{Any, Any}" href="#COBREXA._maybemap-Tuple{Any, Any}"><code>COBREXA._maybemap</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_maybemap(f, x) -&gt; Any
</code></pre><p>Apply a function to <code>x</code> only if it is not <code>nothing</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/Maybe.jl#L18">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Annotations" href="#COBREXA.Annotations"><code>COBREXA.Annotations</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Annotations = Dict{String,Vector{String}}</code></pre><p>Dictionary used to store (possible multiple) standardized annotations of something, such as a <a href="#COBREXA.Metabolite"><code>Metabolite</code></a> and a <a href="#COBREXA.Reaction"><code>Reaction</code></a>.</p><p><strong>Example</strong></p><pre><code class="language-none">Annotations(&quot;PubChem&quot; =&gt; [&quot;CID12345&quot;, &quot;CID54321&quot;])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/MetabolicModel.jl#L47-L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.GeneAssociation" href="#COBREXA.GeneAssociation"><code>COBREXA.GeneAssociation</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">GeneAssociation = Vector{Vector{String}}</code></pre><p>An association to genes, represented as a logical formula in a positive disjunctive normal form (DNF). (The 2nd-level vectors of strings are connected by &quot;and&quot; to form conjunctions, and the 1st-level vectors of these conjunctions are connected by &quot;or&quot; to form the DNF.)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/MetabolicModel.jl#L30-L37">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.MetabolicModel" href="#COBREXA.MetabolicModel"><code>COBREXA.MetabolicModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">abstract type MetabolicModel end</code></pre><p>A helper supertype of everything usable as a linear-like model for COBREXA functions.</p><p>If you want your model type to work with COBREXA, add the <code>MetabolicModel</code> as its supertype, and implement the accessor functions. Accessors <a href="#COBREXA.reactions-Tuple{CoreModel}"><code>reactions</code></a>, <a href="#COBREXA.metabolites-Tuple{CoreModel}"><code>metabolites</code></a>, <a href="#COBREXA.stoichiometry-Tuple{CoreModel}"><code>stoichiometry</code></a>, <a href="#COBREXA.bounds-Tuple{CoreModel}"><code>bounds</code></a> and <a href="#COBREXA.objective-Tuple{CoreModel}"><code>objective</code></a> must be implemented; others are not mandatory and default to safe &quot;empty&quot; values.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/MetabolicModel.jl#L2-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.MetaboliteFormula" href="#COBREXA.MetaboliteFormula"><code>COBREXA.MetaboliteFormula</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">MetaboliteFormula = Dict{String,Int}</code></pre><p>Dictionary of atoms and their abundances in a molecule.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/MetabolicModel.jl#L40-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.ModelWrapper" href="#COBREXA.ModelWrapper"><code>COBREXA.ModelWrapper</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">abstract type ModelWrapper &lt;: MetabolicModel end</code></pre><p>A helper supertype of all &quot;wrapper&quot; types that contain precisely one other <a href="#COBREXA.MetabolicModel"><code>MetabolicModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/MetabolicModel.jl#L16-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Notes" href="#COBREXA.Notes"><code>COBREXA.Notes</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Notes = Dict{String,Vector{String}}</code></pre><p>Free-form notes about something (e.g. a <a href="#COBREXA.Gene"><code>Gene</code></a>), categorized by &quot;topic&quot;.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/abstract/MetabolicModel.jl#L60-L65">source</a></section></article><h2 id="Model-types-and-contents"><a class="docs-heading-anchor" href="#Model-types-and-contents">Model types and contents</a><a id="Model-types-and-contents-1"></a><a class="docs-heading-anchor-permalink" href="#Model-types-and-contents" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.CoreModel" href="#COBREXA.CoreModel"><code>COBREXA.CoreModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct CoreModel &lt;: MetabolicModel</code></pre><p>A &quot;bare bones&quot; core linear optimization problem of the form, with reaction and metabolite names.</p><pre><code class="language-none">min c^T x
s.t. S x = b
      xₗ ≤ x ≤ xᵤ</code></pre><p><strong>Fields</strong></p><ul><li><p><code>S::SparseArrays.SparseMatrixCSC{Float64, Int64}</code></p></li><li><p><code>b::SparseArrays.SparseVector{Float64, Int64}</code></p></li><li><p><code>c::SparseArrays.SparseVector{Float64, Int64}</code></p></li><li><p><code>xl::Vector{Float64}</code></p></li><li><p><code>xu::Vector{Float64}</code></p></li><li><p><code>rxns::Vector{String}</code></p></li><li><p><code>mets::Vector{String}</code></p></li><li><p><code>grrs::Vector{Union{Nothing, Vector{Vector{String}}}}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L2">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Union{Tuple{M}, Tuple{Type{CoreModel}, M}} where M&lt;:MetabolicModel" href="#Base.convert-Union{Tuple{M}, Tuple{Type{CoreModel}, M}} where M&lt;:MetabolicModel"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">convert(_::Type{CoreModel}, m::MetabolicModel) -&gt; MetabolicModel
</code></pre><p>Make a <code>CoreModel</code> out of any compatible model type.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L153">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{CoreModel}" href="#COBREXA.balance-Tuple{CoreModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(a::CoreModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p><code>CoreModel</code> target flux balance.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L78">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{CoreModel}" href="#COBREXA.bounds-Tuple{CoreModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(a::CoreModel) -&gt; Tuple{Vector{Float64}, Vector{Float64}}
</code></pre><p><code>CoreModel</code> flux bounds.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L71">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{CoreModel}" href="#COBREXA.genes-Tuple{CoreModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(a::CoreModel) -&gt; Vector{String}
</code></pre><p>Collect all genes contained in the <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>. The call is expensive for large models, because the vector is not stored and instead gets rebuilt each time this function is called.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L92">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{CoreModel}" href="#COBREXA.metabolites-Tuple{CoreModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(a::CoreModel) -&gt; Vector{String}
</code></pre><p>Metabolites in a <code>CoreModel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{CoreModel}" href="#COBREXA.objective-Tuple{CoreModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(a::CoreModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p><code>CoreModel</code> objective vector.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{CoreModel, Int64}" href="#COBREXA.reaction_gene_association-Tuple{CoreModel, Int64}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(model::CoreModel, ridx::Int64) -&gt; Union{Nothing, Vector{Vector{String}}}
</code></pre><p>Retrieve the <a href="#COBREXA.GeneAssociation"><code>GeneAssociation</code></a> from <a href="#COBREXA.CoreModel"><code>CoreModel</code></a> by reaction index.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L136">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{CoreModel, String}" href="#COBREXA.reaction_gene_association-Tuple{CoreModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(model::CoreModel, rid::String) -&gt; Union{Nothing, Vector{Vector{String}}}
</code></pre><p>Retrieve the <a href="#COBREXA.GeneAssociation"><code>GeneAssociation</code></a> from <a href="#COBREXA.CoreModel"><code>CoreModel</code></a> by reaction ID.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L145">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association_vec-Tuple{CoreModel}" href="#COBREXA.reaction_gene_association_vec-Tuple{CoreModel}"><code>COBREXA.reaction_gene_association_vec</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association_vec(model::CoreModel) -&gt; Vector{Union{Nothing, Vector{Vector{String}}}}
</code></pre><p>Retrieve a vector of all gene associations in a <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>, in the same order as <code>reactions(model)</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L128">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{CoreModel, Any}" href="#COBREXA.reaction_stoichiometry-Tuple{CoreModel, Any}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(m::CoreModel, ridx) -&gt; Dict{String, Float64}
</code></pre><p>Return the stoichiometry of reaction at index <code>ridx</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L120">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{CoreModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{CoreModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(m::CoreModel, rid::String) -&gt; Dict{String, Float64}
</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L112">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{CoreModel}" href="#COBREXA.reactions-Tuple{CoreModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(a::CoreModel) -&gt; Vector{String}
</code></pre><p>Get the reactions in a <code>CoreModel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{CoreModel}" href="#COBREXA.stoichiometry-Tuple{CoreModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(a::CoreModel) -&gt; SparseArrays.SparseMatrixCSC{Float64, Int64}
</code></pre><p><code>CoreModel</code> stoichiometry matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModel.jl#L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.CoreCoupling" href="#COBREXA.CoreCoupling"><code>COBREXA.CoreCoupling</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct CoreCoupling{M} &lt;: ModelWrapper</code></pre><p>A matrix-based wrap that adds reaction coupling matrix to the inner model. A flux <code>x</code> feasible in this model must satisfy:</p><pre><code class="language-none">    cₗ ≤ C x ≤ cᵤ</code></pre><p><strong>Fields</strong></p><ul><li><p><code>lm::Any</code></p></li><li><p><code>C::SparseArrays.SparseMatrixCSC{Float64, Int64}</code></p></li><li><p><code>cl::Vector{Float64}</code></p></li><li><p><code>cu::Vector{Float64}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L2">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.CoreModelCoupled" href="#COBREXA.CoreModelCoupled"><code>COBREXA.CoreModelCoupled</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">const CoreModelCoupled = CoreCoupling{CoreModel}</code></pre><p>A matrix-based linear model with additional coupling constraints in the form:</p><pre><code class="language-none">    cₗ ≤ C x ≤ cᵤ</code></pre><p>Internally, the model is implemented using <a href="#COBREXA.CoreCoupling"><code>CoreCoupling</code></a> that contains a single <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L86-L95">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Union{Tuple{M}, Tuple{Type{CoreCoupling{M}}, MetabolicModel}} where M" href="#Base.convert-Union{Tuple{M}, Tuple{Type{CoreCoupling{M}}, MetabolicModel}} where M"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">convert(::Type{CoreCoupling{M}}, mm::MetabolicModel; clone_coupling) -&gt; MetabolicModel
</code></pre><p>Make a <code>CoreCoupling</code> out of any compatible model type.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{CoreCoupling}" href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(a::CoreCoupling) -&gt; SparseArrays.SparseMatrixCSC{Float64, Int64}
</code></pre><p>Coupling constraint matrix for a <code>CoreCoupling</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{CoreCoupling}" href="#COBREXA.coupling_bounds-Tuple{CoreCoupling}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(a::CoreCoupling) -&gt; Tuple{Vector{Float64}, Vector{Float64}}
</code></pre><p>Coupling bounds for a <code>CoreCoupling</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L56">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_coupling_constraints-Tuple{CoreCoupling}" href="#COBREXA.n_coupling_constraints-Tuple{CoreCoupling}"><code>COBREXA.n_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_coupling_constraints(a::CoreCoupling) -&gt; Int64
</code></pre><p>The number of coupling constraints in a <code>CoreCoupling</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L49">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association_vec-Tuple{CoreModelCoupled}" href="#COBREXA.reaction_gene_association_vec-Tuple{CoreModelCoupled}"><code>COBREXA.reaction_gene_association_vec</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association_vec(model::CoreModelCoupled)</code></pre><p>Evaluates <a href="#COBREXA.reaction_gene_association_vec-Tuple{CoreModel}"><code>reaction_gene_association_vec</code></a> on the model contained in CoreModelCoupled.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L102-L106">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{CoreModelCoupled, Int64}" href="#COBREXA.reaction_stoichiometry-Tuple{CoreModelCoupled, Int64}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::CoreModelCoupled, ridx::Int)</code></pre><p>Evaluates <a href="#COBREXA.reaction_stoichiometry-Tuple{CoreModel, Any}"><code>reaction_stoichiometry</code></a> on the model contained in CoreModelCoupled.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L103-L107">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.unwrap_model-Tuple{CoreCoupling}" href="#COBREXA.unwrap_model-Tuple{CoreCoupling}"><code>COBREXA.unwrap_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">unwrap_model(a::CoreCoupling) -&gt; Any
</code></pre><p>Get the internal <a href="#COBREXA.CoreModel"><code>CoreModel</code></a> out of <a href="#COBREXA.CoreCoupling"><code>CoreCoupling</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/CoreModelCoupled.jl#L35">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.FluxSummary" href="#COBREXA.FluxSummary"><code>COBREXA.FluxSummary</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct FluxSummary</code></pre><p>A struct used to store summary information about the solution of a constraint based analysis result.</p><p><strong>Fields</strong></p><ul><li><p><code>biomass_fluxes::OrderedCollections.OrderedDict{String, Float64}</code></p></li><li><p><code>import_fluxes::OrderedCollections.OrderedDict{String, Float64}</code></p></li><li><p><code>export_fluxes::OrderedCollections.OrderedDict{String, Float64}</code></p></li><li><p><code>unbounded_fluxes::OrderedCollections.OrderedDict{String, Float64}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/FluxSummary.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.FluxSummary-Tuple{}" href="#COBREXA.FluxSummary-Tuple{}"><code>COBREXA.FluxSummary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">FluxSummary() -&gt; FluxSummary
</code></pre><p>A default empty constructor for <code>FluxSummary</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/FluxSummary.jl#L17">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_summary-Tuple{Union{Nothing, Dict{String, Float64}}}" href="#COBREXA.flux_summary-Tuple{Union{Nothing, Dict{String, Float64}}}"><code>COBREXA.flux_summary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_summary(flux_result::Union{Nothing, Dict{String, Float64}}; exclude_exchanges, exchange_prefixes, biomass_strings, exclude_biomass, small_flux_bound, large_flux_bound, keep_unbounded) -&gt; FluxSummary
</code></pre><p>Summarize a dictionary of fluxes into small, useful representation of the most important information contained. Useful for pretty-printing and quickly exploring the results. Internally this function uses <a href="../utils/#COBREXA.looks_like_biomass_reaction-Tuple{String}"><code>looks_like_biomass_reaction</code></a> and <a href="../utils/#COBREXA.looks_like_exchange_reaction-Tuple{String}"><code>looks_like_exchange_reaction</code></a>. The corresponding keyword arguments passed to these functions. Use this if your model has non-standard ids for reactions. Fluxes smaller than <code>small_flux_bound</code> are not stored, while fluxes larger than <code>large_flux_bound</code> are only stored if <code>keep_unbounded</code> is <code>true</code>.</p><p><strong>Example</strong></p><pre><code class="language-none">julia&gt; sol = flux_dict(flux_balance_analysis(model, Tulip.Optimizer))
julia&gt; fr = flux_summary(sol)
Biomass:
  BIOMASS_Ecoli_core_w_GAM: 0.8739
Import:
  EX_o2_e:     -21.7995
  EX_glc__D_e: -10.0
  EX_nh4_e:    -4.7653
  EX_pi_e:     -3.2149
Export:
  EX_h_e:      17.5309
  EX_co2_e:    22.8098
  EX_h2o_e:    29.1758</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/FluxSummary.jl#L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.FluxVariabilitySummary" href="#COBREXA.FluxVariabilitySummary"><code>COBREXA.FluxVariabilitySummary</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct FluxVariabilitySummary</code></pre><p>Stores summary information about the result of a flux variability analysis.</p><p><strong>Fields</strong></p><ul><li><p><code>biomass_fluxes::Dict{String, Vector{Union{Nothing, Float64}}}</code></p></li><li><p><code>exchange_fluxes::Dict{String, Vector{Union{Nothing, Float64}}}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/FluxVariabilitySummary.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.FluxVariabilitySummary-Tuple{}" href="#COBREXA.FluxVariabilitySummary-Tuple{}"><code>COBREXA.FluxVariabilitySummary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">FluxVariabilitySummary() -&gt; FluxVariabilitySummary
</code></pre><p>A default empty constructor for <a href="#COBREXA.FluxVariabilitySummary"><code>FluxVariabilitySummary</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/FluxVariabilitySummary.jl#L14">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_variability_summary-Tuple{Tuple{Dict{String, Dict{String, Float64}}, Dict{String, Dict{String, Float64}}}}" href="#COBREXA.flux_variability_summary-Tuple{Tuple{Dict{String, Dict{String, Float64}}, Dict{String, Dict{String, Float64}}}}"><code>COBREXA.flux_variability_summary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_variability_summary(flux_result::Tuple{Dict{String, Dict{String, Float64}}, Dict{String, Dict{String, Float64}}}; exclude_exchanges, exchange_prefixes, biomass_strings, exclude_biomass) -&gt; FluxVariabilitySummary
</code></pre><p>Summarize a dictionary of flux dictionaries obtained eg. from <a href="../analysis/#COBREXA.flux_variability_analysis_dict-Tuple{MetabolicModel, Any}"><code>flux_variability_analysis_dict</code></a>. The simplified summary representation is useful for pretty-printing and easily showing the most important results.</p><p>Internally this function uses <a href="../utils/#COBREXA.looks_like_biomass_reaction-Tuple{String}"><code>looks_like_biomass_reaction</code></a> and <a href="../utils/#COBREXA.looks_like_exchange_reaction-Tuple{String}"><code>looks_like_exchange_reaction</code></a>. The corresponding keyword arguments are passed to these functions. Use this if your model has an uncommon naming of reactions.</p><p><strong>Example</strong></p><pre><code class="language-none">julia&gt; sol = flux_variability_analysis_dict(model, Gurobi.Optimizer; bounds = objective_bounds(0.99))
julia&gt; flux_res = flux_variability_summary(sol)
Biomass                     Lower bound   Upper bound
  BIOMASS_Ecoli_core_w_GAM: 0.8652        0.8652
Exchange
  EX_h2o_e:                 28.34         28.34
  EX_co2_e:                 22.0377       22.0377
  EX_o2_e:                  -22.1815      -22.1815
  EX_h_e:                   17.3556       17.3556
  EX_glc__D_e:              -10.0         -10.0
  EX_nh4_e:                 -4.8448       -4.8448
  EX_pi_e:                  -3.2149       -3.2149
  EX_for_e:                 0.0           0.0
  ...                       ...           ...</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/FluxVariabilitySummary.jl#L26">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Gene" href="#COBREXA.Gene"><code>COBREXA.Gene</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Gene() -&gt; Gene
Gene(id; name, notes, annotations) -&gt; Gene
</code></pre><p>A convenient constructor for a <code>Gene</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Gene.jl#L14">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Gene" href="#COBREXA.Gene"><code>COBREXA.Gene</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct Gene</code></pre><p><strong>Fields</strong></p><ul><li><p><code>id::String</code></p></li><li><p><code>name::Union{Nothing, String}</code></p></li><li><p><code>notes::Dict{String, Vector{String}}</code></p></li><li><p><code>annotations::Dict{String, Vector{String}}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Gene.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.HDF5Model" href="#COBREXA.HDF5Model"><code>COBREXA.HDF5Model</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct HDF5Model &lt;: MetabolicModel</code></pre><p>A model that is stored in HDF5 format. The model data is never really pulled into memory, but instead mmap&#39;ed as directly as possible into the Julia structures.  This makes reading the <code>HDF5Model</code>s extremely fast, at the same time the (uncached) <code>HDF5Model</code>s can be sent around efficiently among distributed nodes just like <a href="#COBREXA.Serialized"><code>Serialized</code></a> models, provided the nodes share a common storage.</p><p>All HDF5Models must have the backing disk storage. To create one, use <a href="../io/#COBREXA.save_h5_model-Tuple{MetabolicModel, String}"><code>save_h5_model</code></a> or <a href="../io/#COBREXA.save_model-Tuple{MetabolicModel, String}"><code>save_model</code></a> with <code>.h5</code> file extension. To create a temporary model that behaves like a model &quot;in memory&quot;, save it to a temporary file. For related reasons, you can not use <code>convert</code> models to <code>HDF5Model</code> format, because the conversion would impliy having the model saved somewhere.</p><p><strong>Fields</strong></p><ul><li><p><code>h5::Union{Nothing, HDF5.File}</code></p></li><li><p><code>filename::String</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/HDF5Model.jl#L2">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Isozyme" href="#COBREXA.Isozyme"><code>COBREXA.Isozyme</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct Isozyme</code></pre><p>Information about isozyme composition and activity.</p><p><strong>Fields</strong></p><ul><li><p><code>gene_product_count::Dict{String, Int64}</code></p></li><li><p><code>kcat_forward::Float64</code></p></li><li><p><code>kcat_reverse::Float64</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Isozyme.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.JSONModel" href="#COBREXA.JSONModel"><code>COBREXA.JSONModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct JSONModel &lt;: MetabolicModel</code></pre><p>A struct used to store the contents of a JSON model, i.e. a model read from a file ending with <code>.json</code>. These model files typically store all the model data in arrays of JSON objects (represented in Julia as vectors of dictionaries).</p><p>Usually, not all of the fields of the input JSON can be easily represented when converting to other models, care should be taken to avoid losing information.</p><p>Direct work with the <code>json</code> structure is not very efficient; the model structure therefore caches some of the internal structure in the extra fields. The single-parameter <a href="#COBREXA.JSONModel"><code>JSONModel</code></a> constructor creates these caches correctly from the <code>json</code>. The model structure is designed as read-only, and changes in <code>json</code> invalidate the cache.</p><p><strong>Example</strong></p><pre><code class="language-none">model = load_json_model(&quot;some_model.json&quot;)
model.json # see the actual underlying JSON
reactions(model) # see the list of reactions</code></pre><p><strong>Fields</strong></p><ul><li><p><code>json::Dict{String, Any}</code></p></li><li><p><code>rxn_index::Dict{String, Int64}</code></p></li><li><p><code>rxns::Vector{Any}</code></p></li><li><p><code>met_index::Dict{String, Int64}</code></p></li><li><p><code>mets::Vector{Any}</code></p></li><li><p><code>gene_index::Dict{String, Int64}</code></p></li><li><p><code>genes::Vector{Any}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{JSONModel}, MetabolicModel}" href="#Base.convert-Tuple{Type{JSONModel}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">convert(_::Type{JSONModel}, mm::MetabolicModel) -&gt; MetabolicModel
</code></pre><p>Convert any <a href="#COBREXA.MetabolicModel"><code>MetabolicModel</code></a> to <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L284">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{JSONModel}" href="#COBREXA.bounds-Tuple{JSONModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::JSONModel) -&gt; Tuple{Vector, Vector}
</code></pre><p>Get the bounds for reactions, assuming the information is stored in <code>.lower_bound</code> and <code>.upper_bound</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L137">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_annotations-Tuple{JSONModel, String}" href="#COBREXA.gene_annotations-Tuple{JSONModel, String}"><code>COBREXA.gene_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_annotations(model::JSONModel, gid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Gene annotations from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L198">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_name-Tuple{JSONModel, String}" href="#COBREXA.gene_name-Tuple{JSONModel, String}"><code>COBREXA.gene_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_name(model::JSONModel, gid::String) -&gt; Any
</code></pre><p>Return the name of gene with ID <code>gid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L276">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_notes-Tuple{JSONModel, String}" href="#COBREXA.gene_notes-Tuple{JSONModel, String}"><code>COBREXA.gene_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_notes(model::JSONModel, gid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Gene notes from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L208">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{JSONModel}" href="#COBREXA.genes-Tuple{JSONModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(model::JSONModel) -&gt; Vector
</code></pre><p>Extract gene names from a JSON model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L88">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_annotations-Tuple{JSONModel, String}" href="#COBREXA.metabolite_annotations-Tuple{JSONModel, String}"><code>COBREXA.metabolite_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_annotations(model::JSONModel, mid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Metabolite annotations from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L234">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{JSONModel, String}" href="#COBREXA.metabolite_charge-Tuple{JSONModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::JSONModel, mid::String) -&gt; Any
</code></pre><p>Return the metabolite <code>.charge</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L182">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{JSONModel, String}" href="#COBREXA.metabolite_compartment-Tuple{JSONModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(model::JSONModel, mid::String) -&gt; Any
</code></pre><p>Return the metabolite <code>.compartment</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L190">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{JSONModel, String}" href="#COBREXA.metabolite_formula-Tuple{JSONModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(model::JSONModel, mid::String) -&gt; Union{Nothing, Dict{String, Int64}}
</code></pre><p>Parse and return the metabolite <code>.formula</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L174">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_name-Tuple{JSONModel, String}" href="#COBREXA.metabolite_name-Tuple{JSONModel, String}"><code>COBREXA.metabolite_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_name(model::JSONModel, mid::String) -&gt; Any
</code></pre><p>Return the name of metabolite with ID <code>mid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L268">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_notes-Tuple{JSONModel, String}" href="#COBREXA.metabolite_notes-Tuple{JSONModel, String}"><code>COBREXA.metabolite_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_notes(model::JSONModel, mid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Metabolite notes from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L244">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{JSONModel}" href="#COBREXA.metabolites-Tuple{JSONModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(model::JSONModel) -&gt; Vector
</code></pre><p>Extract metabolite names (stored as <code>.id</code>) from JSON model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L81">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{JSONModel}" href="#COBREXA.objective-Tuple{JSONModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::JSONModel) -&gt; SparseArrays.SparseVector{_A, Int64} where _A
</code></pre><p>Collect <code>.objective_coefficient</code> keys from model reactions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L148">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_annotations-Tuple{JSONModel, String}" href="#COBREXA.reaction_annotations-Tuple{JSONModel, String}"><code>COBREXA.reaction_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_annotations(model::JSONModel, rid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Reaction annotations from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L216">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{JSONModel, String}" href="#COBREXA.reaction_gene_association-Tuple{JSONModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(model::JSONModel, rid::String) -&gt; Union{Nothing, Vector{Vector{String}}}
</code></pre><p>Parses the <code>.gene_reaction_rule</code> from reactions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L156">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_name-Tuple{JSONModel, String}" href="#COBREXA.reaction_name-Tuple{JSONModel, String}"><code>COBREXA.reaction_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_name(model::JSONModel, rid::String) -&gt; Any
</code></pre><p>Return the name of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L260">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_notes-Tuple{JSONModel, String}" href="#COBREXA.reaction_notes-Tuple{JSONModel, String}"><code>COBREXA.reaction_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_notes(model::JSONModel, rid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Reaction notes from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L226">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{JSONModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{JSONModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::JSONModel, rid::String) -&gt; Dict{String, Float64}
</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L252">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_subsystem-Tuple{JSONModel, String}" href="#COBREXA.reaction_subsystem-Tuple{JSONModel, String}"><code>COBREXA.reaction_subsystem</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_subsystem(model::JSONModel, rid::String) -&gt; Any
</code></pre><p>Parses the <code>.subsystem</code> out from reactions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L166">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{JSONModel}" href="#COBREXA.reactions-Tuple{JSONModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::JSONModel) -&gt; Vector
</code></pre><p>Extract reaction names (stored as <code>.id</code>) from JSON model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L74">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{JSONModel}" href="#COBREXA.stoichiometry-Tuple{JSONModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::JSONModel) -&gt; SparseArrays.SparseMatrixCSC{Float64, Int64}
</code></pre><p>Get the stoichiometry. Assuming the information is stored in reaction object under key <code>.metabolites</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/JSONModel.jl#L95">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.MATModel" href="#COBREXA.MATModel"><code>COBREXA.MATModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct MATModel &lt;: MetabolicModel</code></pre><p>Wrapper around the models loaded in dictionaries from the MATLAB representation.</p><p><strong>Fields</strong></p><ul><li><code>mat::Dict{String, Any}</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{MATModel}, MetabolicModel}" href="#Base.convert-Tuple{Type{MATModel}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">convert(_::Type{MATModel}, m::MetabolicModel) -&gt; MetabolicModel
</code></pre><p>Convert any metabolic model to <code>MATModel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L212">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._mat_has_squashed_coupling-Tuple{Any}" href="#COBREXA._mat_has_squashed_coupling-Tuple{Any}"><code>COBREXA._mat_has_squashed_coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_mat_has_squashed_coupling(mat) -&gt; Any
</code></pre><p>Guesses whether C in the MAT file is stored in A=[S;C].</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L29">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{MATModel}" href="#COBREXA.balance-Tuple{MATModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(m::MATModel) -&gt; Any
</code></pre><p>Extracts balance from the MAT model, defaulting to zeroes if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L69">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{MATModel}" href="#COBREXA.bounds-Tuple{MATModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(m::MATModel) -&gt; Tuple{Any, Any}
</code></pre><p>Extracts bounds from the MAT file, saved under <code>lb</code> and <code>ub</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{MATModel}" href="#COBREXA.coupling-Tuple{MATModel}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(m::MATModel) -&gt; Any
</code></pre><p>Extract coupling matrix stored, in <code>C</code> key.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L90">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{MATModel}" href="#COBREXA.coupling_bounds-Tuple{MATModel}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(m::MATModel) -&gt; Tuple{Any, Any}
</code></pre><p>Extracts the coupling constraints. Currently, there are several accepted ways to store these in MATLAB models; this takes the constraints from vectors <code>cl</code> and <code>cu</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L99">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{MATModel}" href="#COBREXA.genes-Tuple{MATModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(m::MATModel) -&gt; Any
</code></pre><p>Extracts the possible gene list from <code>genes</code> key.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L119">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{MATModel, String}" href="#COBREXA.metabolite_charge-Tuple{MATModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(m::MATModel, mid::String) -&gt; Union{Nothing, Int64}
</code></pre><p>Extract metabolite charge from <code>metCharge</code> or <code>metCharges</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L153">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{MATModel, String}" href="#COBREXA.metabolite_compartment-Tuple{MATModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(m::MATModel, mid::String) -&gt; Any
</code></pre><p>Extract metabolite compartment from <code>metCompartment</code> or <code>metCompartments</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L166">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{MATModel, String}" href="#COBREXA.metabolite_formula-Tuple{MATModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(m::MATModel, mid::String) -&gt; Union{Nothing, Dict{String, Int64}}
</code></pre><p>Extract metabolite formula from key <code>metFormula</code> or <code>metFormulas</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L143">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_name-Tuple{MATModel, String}" href="#COBREXA.metabolite_name-Tuple{MATModel, String}"><code>COBREXA.metabolite_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_name(m::MATModel, mid::String) -&gt; Any
</code></pre><p>Extract metabolite name from <code>metNames</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L272">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{MATModel}" href="#COBREXA.metabolites-Tuple{MATModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(m::MATModel) -&gt; Vector{String}
</code></pre><p>Extracts metabolite names from <code>mets</code> key in the MAT file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L38">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{MATModel}" href="#COBREXA.objective-Tuple{MATModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(m::MATModel) -&gt; Any
</code></pre><p>Extracts the objective from the MAT model (defaults to zeroes).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L82">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{MATModel, String}" href="#COBREXA.reaction_gene_association-Tuple{MATModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(m::MATModel, rid::String) -&gt; Union{Nothing, Vector{Vector{String}}}
</code></pre><p>Extracts the associations from <code>grRules</code> key, if present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L129">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_name-Tuple{MATModel, String}" href="#COBREXA.reaction_name-Tuple{MATModel, String}"><code>COBREXA.reaction_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_name(m::MATModel, rid::String) -&gt; Any
</code></pre><p>Extract reaction name from <code>rxnNames</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L262">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{MATModel, Any}" href="#COBREXA.reaction_stoichiometry-Tuple{MATModel, Any}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(m::MATModel, ridx) -&gt; Dict{String, Float64}
</code></pre><p>Return the stoichiometry of reaction at index <code>ridx</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L195">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{MATModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{MATModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(m::MATModel, rid::String) -&gt; Dict{String, Float64}
</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L185">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{MATModel}" href="#COBREXA.reactions-Tuple{MATModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(m::MATModel) -&gt; Vector{String}
</code></pre><p>Extracts reaction names from <code>rxns</code> key in the MAT file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{MATModel}" href="#COBREXA.stoichiometry-Tuple{MATModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(m::MATModel) -&gt; Any
</code></pre><p>Extract the stoichiometry matrix, stored under key <code>S</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MATModel.jl#L52">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{MetabolicModel}" href="#COBREXA.balance-Tuple{MetabolicModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(a::MetabolicModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p>Get the sparse balance vector of a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L83">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{MetabolicModel}" href="#COBREXA.bounds-Tuple{MetabolicModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(a::MetabolicModel) -&gt; Tuple{Vector{Float64}, Vector{Float64}}
</code></pre><p>Get the lower and upper solution bounds of a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L74">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{MetabolicModel}" href="#COBREXA.coupling-Tuple{MetabolicModel}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(a::MetabolicModel) -&gt; SparseArrays.SparseMatrixCSC{Float64, Int64}
</code></pre><p>Get a matrix of coupling constraint definitions of a model. By default, there is no coupling in the models.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L138">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{MetabolicModel}" href="#COBREXA.coupling_bounds-Tuple{MetabolicModel}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(a::MetabolicModel) -&gt; Tuple{Vector{Float64}, Vector{Float64}}
</code></pre><p>Get the lower and upper bounds for each coupling bound in a model, as specified by <code>coupling</code>. By default, the model does not have any coupling bounds.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L157">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.fluxes-Tuple{MetabolicModel}" href="#COBREXA.fluxes-Tuple{MetabolicModel}"><code>COBREXA.fluxes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">fluxes(a::MetabolicModel) -&gt; Vector{String}
</code></pre><p>In some models, the <a href="#COBREXA.reactions-Tuple{CoreModel}"><code>reactions</code></a> that correspond to the columns of <a href="#COBREXA.stoichiometry-Tuple{CoreModel}"><code>stoichiometry</code></a> matrix do not fully represent the semantic contents of the model; for example, fluxes may be split into forward and reverse reactions, reactions catalyzed by distinct enzymes, etc. Together with <a href="#COBREXA.reaction_flux-Tuple{MetabolicModel}"><code>reaction_flux</code></a> (and <a href="#COBREXA.n_fluxes-Tuple{ModelWrapper}"><code>n_fluxes</code></a>) this specifies how the flux is decomposed into individual reactions.</p><p>By default (and in most models), fluxes and reactions perfectly correspond.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L103">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_annotations-Tuple{MetabolicModel, String}" href="#COBREXA.gene_annotations-Tuple{MetabolicModel, String}"><code>COBREXA.gene_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_annotations(a::MetabolicModel, gene_id::String) -&gt; Any
</code></pre><p>Return standardized names that identify the corresponding gene or product. The dictionary assigns vectors of possible identifiers to identifier system names, e.g. <code>&quot;PDB&quot; =&gt; [&quot;PROT01&quot;]</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L286">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_name-Tuple{MetabolicModel, String}" href="#COBREXA.gene_name-Tuple{MetabolicModel, String}"><code>COBREXA.gene_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_name(model::MetabolicModel, gid::String) -&gt; Any
</code></pre><p>Return the name of gene with ID <code>gid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L338">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_notes-Tuple{MetabolicModel, String}" href="#COBREXA.gene_notes-Tuple{MetabolicModel, String}"><code>COBREXA.gene_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_notes(model::MetabolicModel, gene_id::String) -&gt; Any
</code></pre><p>Return the notes associated with the gene <code>gene_id</code> in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L315">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{MetabolicModel}" href="#COBREXA.genes-Tuple{MetabolicModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(a::MetabolicModel) -&gt; Vector{String}
</code></pre><p>Return identifiers of all genes contained in the model. By default, there are no genes.</p><p>In SBML, these are usually called &quot;gene products&quot; but we write <code>genes</code> for simplicity.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L167">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_annotations-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_annotations-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_annotations(a::MetabolicModel, metabolite_id::String) -&gt; Any
</code></pre><p>Return standardized names that may help to reliably identify the metabolite. The dictionary assigns vectors of possible identifiers to identifier system names, e.g. <code>&quot;ChEMBL&quot; =&gt; [&quot;123&quot;]</code> or <code>&quot;PubChem&quot; =&gt; [&quot;CID123&quot;, &quot;CID654645645&quot;]</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L275">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_charge-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::MetabolicModel, metabolite_id::String) -&gt; Any
</code></pre><p>Return the charge associated with metabolite <code>metabolite_id</code> in <code>model</code>. Returns <code>nothing</code> if charge not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L244">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_compartment-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(model::MetabolicModel, metabolite_id::String) -&gt; Any
</code></pre><p>Return the compartment of metabolite <code>metabolite_id</code> in <code>model</code> if it is assigned. If not, return <code>nothing</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L254">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_formula-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(model::MetabolicModel, metabolite_id::String) -&gt; Any
</code></pre><p>Return the formula of metabolite <code>metabolite_id</code> in <code>model</code>. Return <code>nothing</code> in case the formula is not known or irrelevant.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L231">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_name-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_name-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_name(model::MetabolicModel, mid::String) -&gt; Any
</code></pre><p>Return the name of metabolite with ID <code>mid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L331">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_notes-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_notes-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_notes(model::MetabolicModel, metabolite_id::String) -&gt; Any
</code></pre><p>Return the notes associated with metabolite <code>reaction_id</code> in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L306">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{MetabolicModel}" href="#COBREXA.metabolites-Tuple{MetabolicModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(a::MetabolicModel) -&gt; Vector{String}
</code></pre><p>Return a vector of metabolite identifiers in a model. The vector precisely corresponds to the rows in <a href="#COBREXA.stoichiometry-Tuple{CoreModel}"><code>stoichiometry</code></a> matrix.</p><p>As with <a href="#COBREXA.reactions-Tuple{CoreModel}"><code>reactions</code></a>s, some metabolites in models may be virtual, representing purely technical equality constraints.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L28">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_coupling_constraints-Tuple{MetabolicModel}" href="#COBREXA.n_coupling_constraints-Tuple{MetabolicModel}"><code>COBREXA.n_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_coupling_constraints(a::MetabolicModel) -&gt; Int64
</code></pre><p>Get the number of coupling constraints in a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L148">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_genes-Tuple{MetabolicModel}" href="#COBREXA.n_genes-Tuple{MetabolicModel}"><code>COBREXA.n_genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_genes(a::MetabolicModel) -&gt; Int64
</code></pre><p>Return the number of genes in the model (as returned by <a href="#COBREXA.genes-Tuple{CoreModel}"><code>genes</code></a>). If you just need the number of the genes, this may be much more efficient than calling <a href="#COBREXA.genes-Tuple{CoreModel}"><code>genes</code></a> and measuring the array.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L180">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_metabolites-Tuple{MetabolicModel}" href="#COBREXA.n_metabolites-Tuple{MetabolicModel}"><code>COBREXA.n_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_metabolites(a::MetabolicModel) -&gt; Int64
</code></pre><p>Get the number of metabolites in a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_reactions-Tuple{MetabolicModel}" href="#COBREXA.n_reactions-Tuple{MetabolicModel}"><code>COBREXA.n_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_reactions(a::MetabolicModel) -&gt; Int64
</code></pre><p>Get the number of reactions in a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L41">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{MetabolicModel}" href="#COBREXA.objective-Tuple{MetabolicModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(a::MetabolicModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p>Get the objective vector of the model. Analysis functions, such as <a href="../analysis/#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>flux_balance_analysis</code></a>, are supposed to maximize <code>dot(objective, x)</code> where <code>x</code> is a feasible solution of the model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L92">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.precache!-Tuple{MetabolicModel}" href="#COBREXA.precache!-Tuple{MetabolicModel}"><code>COBREXA.precache!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">precache!(a::MetabolicModel)
</code></pre><p>Do whatever is feasible to get the model into a state that can be read from as-quickly-as-possible. This may include e.g. generating helper index structures and loading delayed parts of the model from disk. The model should be modified &quot;transparently&quot; in-place. Analysis functions call this right before applying modifications or converting the model to the optimization model using <a href="../base/#COBREXA.make_optimization_model-Tuple{MetabolicModel, Any}"><code>make_optimization_model</code></a>; usually on the same machine where the optimizers (and, generally, the core analysis algorithms) will run. The calls are done in a good hope that the performance will be improved.</p><p>By default, it should be safe to do nothing.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L345">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_annotations-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_annotations-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_annotations(a::MetabolicModel, reaction_id::String) -&gt; Any
</code></pre><p>Return standardized names that may help identifying the reaction. The dictionary assigns vectors of possible identifiers to identifier system names, e.g. <code>&quot;Reactome&quot; =&gt; [&quot;reactomeID123&quot;]</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L264">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_flux-Tuple{MetabolicModel}" href="#COBREXA.reaction_flux-Tuple{MetabolicModel}"><code>COBREXA.reaction_flux</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_flux(a::MetabolicModel) -&gt; Any
</code></pre><p>Retrieve a sparse matrix that describes the correspondence of a solution of the linear system to the fluxes (see <a href="#COBREXA.fluxes-Tuple{MetabolicModel}"><code>fluxes</code></a> for rationale). Returns a sparse matrix of size <code>(n_reactions(a), n_fluxes(a))</code>. For most models, this is an identity matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L123">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_gene_association-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(a::MetabolicModel, reaction_id::String) -&gt; Any
</code></pre><p>Returns the sets of genes that need to be present so that the reaction can work (technically, a DNF on gene availability, with positive atoms only).</p><p>For simplicity, <code>nothing</code> may be returned, meaning that the reaction always takes place. (in DNF, that would be equivalent to returning <code>[[]]</code>.)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L191">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_name-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_name-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_name(model::MetabolicModel, rid::String) -&gt; Any
</code></pre><p>Return the name of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L324">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_notes-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_notes-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_notes(model::MetabolicModel, reaction_id::String) -&gt; Any
</code></pre><p>Return the notes associated with reaction <code>reaction_id</code> in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L297">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(m::MetabolicModel, rid::String) -&gt; Any
</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code> in the model. The dictionary maps the metabolite IDs to their stoichiometric coefficients.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L217">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_subsystem-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_subsystem-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_subsystem</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_subsystem(model::MetabolicModel, reaction_id::String) -&gt; Any
</code></pre><p>Return the subsystem of reaction <code>reaction_id</code> in <code>model</code> if it is assigned. If not, return <code>nothing</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L207">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{MetabolicModel}" href="#COBREXA.reactions-Tuple{MetabolicModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(a::MetabolicModel) -&gt; Vector{String}
</code></pre><p>Return a vector of reaction identifiers in a model. The vector precisely corresponds to the columns in <a href="#COBREXA.stoichiometry-Tuple{CoreModel}"><code>stoichiometry</code></a> matrix.</p><p>For technical reasons, the &quot;reactions&quot; may sometimes not be true reactions but various virtual and helper pseudo-reactions that are used in the metabolic modeling, such as metabolite exchanges, separate forward and reverse reactions, supplies of enzymatic and genetic material and virtual cell volume, etc. To simplify the view of the model contents use <a href="#COBREXA.reaction_flux-Tuple{MetabolicModel}"><code>reaction_flux</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L12">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{MetabolicModel}" href="#COBREXA.stoichiometry-Tuple{MetabolicModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(a::MetabolicModel) -&gt; SparseArrays.SparseMatrixCSC{Float64, Int64}
</code></pre><p>Get the sparse stoichiometry matrix of a model. A feasible solution <code>x</code> of a model <code>m</code> is defined as satisfying the equations:</p><ul><li><code>stoichiometry(m) * x .== balance(m)</code></li><li><code>x .&gt;= lbs</code></li><li><code>y .&lt;= ubs</code></li><li>`(lbs, ubs) == bounds(m)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/MetabolicModel.jl#L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Metabolite" href="#COBREXA.Metabolite"><code>COBREXA.Metabolite</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Metabolite() -&gt; Metabolite
Metabolite(id; name, formula, charge, compartment, notes, annotations) -&gt; Metabolite
</code></pre><p>A constructor for <code>Metabolite</code>s.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Metabolite.jl#L17">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Metabolite" href="#COBREXA.Metabolite"><code>COBREXA.Metabolite</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct Metabolite</code></pre><p><strong>Fields</strong></p><ul><li><p><code>id::String</code></p></li><li><p><code>name::Union{Nothing, String}</code></p></li><li><p><code>formula::Union{Nothing, String}</code></p></li><li><p><code>charge::Union{Nothing, Int64}</code></p></li><li><p><code>compartment::Union{Nothing, String}</code></p></li><li><p><code>notes::Dict{String, Vector{String}}</code></p></li><li><p><code>annotations::Dict{String, Vector{String}}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Metabolite.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{ModelWrapper}" href="#COBREXA.balance-Tuple{ModelWrapper}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.balance-Tuple{CoreModel}"><code>balance</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{ModelWrapper}" href="#COBREXA.bounds-Tuple{ModelWrapper}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.bounds-Tuple{CoreModel}"><code>bounds</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{ModelWrapper}" href="#COBREXA.coupling-Tuple{ModelWrapper}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>coupling</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{ModelWrapper}" href="#COBREXA.coupling_bounds-Tuple{ModelWrapper}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.coupling_bounds-Tuple{CoreCoupling}"><code>coupling_bounds</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.fluxes-Tuple{ModelWrapper}" href="#COBREXA.fluxes-Tuple{ModelWrapper}"><code>COBREXA.fluxes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">fluxes(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.fluxes-Tuple{MetabolicModel}"><code>fluxes</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_annotations-Tuple{ModelWrapper, String}" href="#COBREXA.gene_annotations-Tuple{ModelWrapper, String}"><code>COBREXA.gene_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_annotations(model::ModelWrapper, gid::String)</code></pre><p>Evaluates <a href="#COBREXA.gene_annotations-Tuple{JSONModel, String}"><code>gene_annotations</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L23-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_notes-Tuple{ModelWrapper, String}" href="#COBREXA.gene_notes-Tuple{ModelWrapper, String}"><code>COBREXA.gene_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_notes(model::ModelWrapper, gid::String)</code></pre><p>Evaluates <a href="#COBREXA.gene_notes-Tuple{JSONModel, String}"><code>gene_notes</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L23-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{ModelWrapper}" href="#COBREXA.genes-Tuple{ModelWrapper}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.genes-Tuple{CoreModel}"><code>genes</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_annotations-Tuple{ModelWrapper, String}" href="#COBREXA.metabolite_annotations-Tuple{ModelWrapper, String}"><code>COBREXA.metabolite_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_annotations(model::ModelWrapper, mid::String)</code></pre><p>Evaluates <a href="#COBREXA.metabolite_annotations-Tuple{JSONModel, String}"><code>metabolite_annotations</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L21-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{ModelWrapper, String}" href="#COBREXA.metabolite_charge-Tuple{ModelWrapper, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::ModelWrapper, mid::String)</code></pre><p>Evaluates <a href="#COBREXA.metabolite_charge-Tuple{JSONModel, String}"><code>metabolite_charge</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L21-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{ModelWrapper, String}" href="#COBREXA.metabolite_compartment-Tuple{ModelWrapper, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(model::ModelWrapper, mid::String)</code></pre><p>Evaluates <a href="#COBREXA.metabolite_compartment-Tuple{JSONModel, String}"><code>metabolite_compartment</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L21-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{ModelWrapper, String}" href="#COBREXA.metabolite_formula-Tuple{ModelWrapper, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(model::ModelWrapper, mid::String)</code></pre><p>Evaluates <a href="#COBREXA.metabolite_formula-Tuple{JSONModel, String}"><code>metabolite_formula</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L21-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_notes-Tuple{ModelWrapper, String}" href="#COBREXA.metabolite_notes-Tuple{ModelWrapper, String}"><code>COBREXA.metabolite_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_notes(model::ModelWrapper, mid::String)</code></pre><p>Evaluates <a href="#COBREXA.metabolite_notes-Tuple{JSONModel, String}"><code>metabolite_notes</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L21-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{ModelWrapper}" href="#COBREXA.metabolites-Tuple{ModelWrapper}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.metabolites-Tuple{CoreModel}"><code>metabolites</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_coupling_constraints-Tuple{ModelWrapper}" href="#COBREXA.n_coupling_constraints-Tuple{ModelWrapper}"><code>COBREXA.n_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_coupling_constraints(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.n_coupling_constraints-Tuple{CoreCoupling}"><code>n_coupling_constraints</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_fluxes-Tuple{ModelWrapper}" href="#COBREXA.n_fluxes-Tuple{ModelWrapper}"><code>COBREXA.n_fluxes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_fluxes(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.n_fluxes-Tuple{ModelWrapper}"><code>n_fluxes</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_genes-Tuple{ModelWrapper}" href="#COBREXA.n_genes-Tuple{ModelWrapper}"><code>COBREXA.n_genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_genes(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.n_genes-Tuple{MetabolicModel}"><code>n_genes</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{ModelWrapper}" href="#COBREXA.objective-Tuple{ModelWrapper}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.objective-Tuple{CoreModel}"><code>objective</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.precache!-Tuple{ModelWrapper}" href="#COBREXA.precache!-Tuple{ModelWrapper}"><code>COBREXA.precache!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">precache!(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.precache!-Tuple{MetabolicModel}"><code>precache!</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_annotations-Tuple{ModelWrapper, String}" href="#COBREXA.reaction_annotations-Tuple{ModelWrapper, String}"><code>COBREXA.reaction_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_annotations(model::ModelWrapper, rid::String)</code></pre><p>Evaluates <a href="#COBREXA.reaction_annotations-Tuple{JSONModel, String}"><code>reaction_annotations</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L19-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_flux-Tuple{ModelWrapper}" href="#COBREXA.reaction_flux-Tuple{ModelWrapper}"><code>COBREXA.reaction_flux</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_flux(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.reaction_flux-Tuple{MetabolicModel}"><code>reaction_flux</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{ModelWrapper, String}" href="#COBREXA.reaction_gene_association-Tuple{ModelWrapper, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(model::ModelWrapper, rid::String)</code></pre><p>Evaluates <a href="#COBREXA.reaction_gene_association-Tuple{CoreModel, Int64}"><code>reaction_gene_association</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L19-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_notes-Tuple{ModelWrapper, String}" href="#COBREXA.reaction_notes-Tuple{ModelWrapper, String}"><code>COBREXA.reaction_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_notes(model::ModelWrapper, rid::String)</code></pre><p>Evaluates <a href="#COBREXA.reaction_notes-Tuple{JSONModel, String}"><code>reaction_notes</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L19-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{ModelWrapper, String}" href="#COBREXA.reaction_stoichiometry-Tuple{ModelWrapper, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::ModelWrapper, rid::String)</code></pre><p>Evaluates <a href="#COBREXA.reaction_stoichiometry-Tuple{CoreModel, Any}"><code>reaction_stoichiometry</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L19-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_subsystem-Tuple{ModelWrapper, String}" href="#COBREXA.reaction_subsystem-Tuple{ModelWrapper, String}"><code>COBREXA.reaction_subsystem</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_subsystem(model::ModelWrapper, rid::String)</code></pre><p>Evaluates <a href="#COBREXA.reaction_subsystem-Tuple{JSONModel, String}"><code>reaction_subsystem</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L19-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{ModelWrapper}" href="#COBREXA.reactions-Tuple{ModelWrapper}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.reactions-Tuple{CoreModel}"><code>reactions</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{ModelWrapper}" href="#COBREXA.stoichiometry-Tuple{ModelWrapper}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::ModelWrapper)</code></pre><p>Evaluates <a href="#COBREXA.stoichiometry-Tuple{CoreModel}"><code>stoichiometry</code></a> on the model contained in ModelWrapper.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L17-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.unwrap_model-Tuple{ModelWrapper}" href="#COBREXA.unwrap_model-Tuple{ModelWrapper}"><code>COBREXA.unwrap_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">unwrap_model(a::ModelWrapper) -&gt; Any
</code></pre><p>A simple helper to pick the single w</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ModelWrapper.jl#L2">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Reaction" href="#COBREXA.Reaction"><code>COBREXA.Reaction</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Reaction(id::String, metabolites) -&gt; Reaction
Reaction(id::String, metabolites, dir; default_bound) -&gt; Reaction
</code></pre><p>Convenience constructor for <code>Reaction</code>. The reaction equation is specified using <code>metabolites</code>, which is a dictionary mapping metabolite ids to stoichiometric coefficients. The direcion of the reaction is set through <code>dir</code> which can take <code>:bidirectional</code>, <code>:forward</code>, and <code>:reverse</code> as values. Finally, the <code>default_bound</code> is the value taken to mean infinity in the context of constraint based models, often this is set to a very high flux value like 1000.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Reaction.jl#L56">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Reaction" href="#COBREXA.Reaction"><code>COBREXA.Reaction</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Reaction() -&gt; Reaction
Reaction(id; name, metabolites, lb, ub, grr, subsystem, notes, annotations, objective_coefficient) -&gt; Reaction
</code></pre><p>A constructor for Reaction that only takes a reaction <code>id</code> and assigns default/uninformative values to all the fields that are not explicitely assigned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Reaction.jl#L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Reaction" href="#COBREXA.Reaction"><code>COBREXA.Reaction</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct Reaction</code></pre><p>A structure for representing a single reaction in a <a href="#COBREXA.StandardModel"><code>StandardModel</code></a>.</p><p><strong>Fields</strong></p><ul><li><p><code>id::String</code></p></li><li><p><code>name::Union{Nothing, String}</code></p></li><li><p><code>metabolites::Dict{String, Float64}</code></p></li><li><p><code>lb::Float64</code></p></li><li><p><code>ub::Float64</code></p></li><li><p><code>grr::Union{Nothing, Vector{Vector{String}}}</code></p></li><li><p><code>subsystem::Union{Nothing, String}</code></p></li><li><p><code>notes::Dict{String, Vector{String}}</code></p></li><li><p><code>annotations::Dict{String, Vector{String}}</code></p></li><li><p><code>objective_coefficient::Float64</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Reaction.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.ReactionStatus" href="#COBREXA.ReactionStatus"><code>COBREXA.ReactionStatus</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct ReactionStatus</code></pre><p>Used for concise reporting of modeling results.</p><p><strong>Fields</strong></p><ul><li><p><code>already_present::Bool</code></p></li><li><p><code>index::Int64</code></p></li><li><p><code>info::String</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/ReactionStatus.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.SBMLModel" href="#COBREXA.SBMLModel"><code>COBREXA.SBMLModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct SBMLModel &lt;: MetabolicModel</code></pre><p>Construct the SBML model and add the necessary cached indexes, possibly choosing an active objective.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L20">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.SBMLModel" href="#COBREXA.SBMLModel"><code>COBREXA.SBMLModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct SBMLModel &lt;: MetabolicModel</code></pre><p>Thin wrapper around the model from SBML.jl library. Allows easy conversion from SBML to any other model format.</p><p><strong>Fields</strong></p><ul><li><p><code>sbml::SBML.Model</code></p></li><li><p><code>reaction_ids::Vector{String}</code></p></li><li><p><code>reaction_idx::Dict{String, Int64}</code></p></li><li><p><code>metabolite_ids::Vector{String}</code></p></li><li><p><code>metabolite_idx::Dict{String, Int64}</code></p></li><li><p><code>gene_ids::Vector{String}</code></p></li><li><p><code>active_objective::String</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{SBMLModel}, MetabolicModel}" href="#Base.convert-Tuple{Type{SBMLModel}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">convert(_::Type{SBMLModel}, mm::MetabolicModel) -&gt; MetabolicModel
</code></pre><p>Convert any metabolic model to <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L386">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{SBMLModel}" href="#COBREXA.balance-Tuple{SBMLModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(model::SBMLModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p>Balance vector of a <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>. This is always zero.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L170">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{SBMLModel}" href="#COBREXA.bounds-Tuple{SBMLModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::SBMLModel) -&gt; Tuple{Vector{Float64}, Vector{Float64}}
</code></pre><p>Get the lower and upper flux bounds of model <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>. Throws <code>DomainError</code> in case if the SBML contains mismatching units.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L111">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_annotations-Tuple{SBMLModel, String}" href="#COBREXA.gene_annotations-Tuple{SBMLModel, String}"><code>COBREXA.gene_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_annotations(model::SBMLModel, gid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the annotations of gene with ID <code>gid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L352">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_name-Tuple{SBMLModel, String}" href="#COBREXA.gene_name-Tuple{SBMLModel, String}"><code>COBREXA.gene_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_name(model::SBMLModel, gid::String) -&gt; Union{Nothing, String}
</code></pre><p>Return the name of gene with ID <code>gid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L329">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_notes-Tuple{SBMLModel, String}" href="#COBREXA.gene_notes-Tuple{SBMLModel, String}"><code>COBREXA.gene_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_notes(model::SBMLModel, gid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the notes about gene with ID <code>gid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L378">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{SBMLModel}" href="#COBREXA.genes-Tuple{SBMLModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(model::SBMLModel) -&gt; Vector{String}
</code></pre><p>Get genes of a <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L204">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_annotations-Tuple{SBMLModel, String}" href="#COBREXA.metabolite_annotations-Tuple{SBMLModel, String}"><code>COBREXA.metabolite_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_annotations(model::SBMLModel, mid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the annotations of metabolite with ID <code>mid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L344">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{SBMLModel, String}" href="#COBREXA.metabolite_charge-Tuple{SBMLModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::SBMLModel, mid::String) -&gt; Union{Nothing, Int64}
</code></pre><p>Get charge of a chosen metabolite from <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L241">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{SBMLModel, String}" href="#COBREXA.metabolite_compartment-Tuple{SBMLModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(model::SBMLModel, mid::String) -&gt; String
</code></pre><p>Get the compartment of a chosen metabolite from <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L234">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{SBMLModel, String}" href="#COBREXA.metabolite_formula-Tuple{SBMLModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(model::SBMLModel, mid::String) -&gt; Union{Nothing, Dict{String, Int64}}
</code></pre><p>Get <a href="#COBREXA.MetaboliteFormula"><code>MetaboliteFormula</code></a> from a chosen metabolite from <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L226">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_name-Tuple{SBMLModel, String}" href="#COBREXA.metabolite_name-Tuple{SBMLModel, String}"><code>COBREXA.metabolite_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_name(model::SBMLModel, mid::String) -&gt; Union{Nothing, String}
</code></pre><p>Return the name of metabolite with ID <code>mid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L322">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_notes-Tuple{SBMLModel, String}" href="#COBREXA.metabolite_notes-Tuple{SBMLModel, String}"><code>COBREXA.metabolite_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_notes(model::SBMLModel, mid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the notes about metabolite with ID <code>mid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L370">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{SBMLModel}" href="#COBREXA.metabolites-Tuple{SBMLModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(model::SBMLModel) -&gt; Vector{String}
</code></pre><p>Get metabolites from a <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L48">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_genes-Tuple{SBMLModel}" href="#COBREXA.n_genes-Tuple{SBMLModel}"><code>COBREXA.n_genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_genes(model::SBMLModel) -&gt; Int64
</code></pre><p>Get number of genes in <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L211">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_metabolites-Tuple{SBMLModel}" href="#COBREXA.n_metabolites-Tuple{SBMLModel}"><code>COBREXA.n_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_metabolites(model::SBMLModel) -&gt; Int64
</code></pre><p>Efficient counting of metabolites in <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L62">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_reactions-Tuple{SBMLModel}" href="#COBREXA.n_reactions-Tuple{SBMLModel}"><code>COBREXA.n_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_reactions(model::SBMLModel) -&gt; Int64
</code></pre><p>Efficient counting of reactions in <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L55">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{SBMLModel}" href="#COBREXA.objective-Tuple{SBMLModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::SBMLModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p>Objective of the <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L177">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_annotations-Tuple{SBMLModel, String}" href="#COBREXA.reaction_annotations-Tuple{SBMLModel, String}"><code>COBREXA.reaction_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_annotations(model::SBMLModel, rid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the annotations of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L336">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{SBMLModel, String}" href="#COBREXA.reaction_gene_association-Tuple{SBMLModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(model::SBMLModel, rid::String) -&gt; Union{Nothing, Vector{Vector{String}}}
</code></pre><p>Retrieve the <a href="#COBREXA.GeneAssociation"><code>GeneAssociation</code></a> from <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L218">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_name-Tuple{SBMLModel, String}" href="#COBREXA.reaction_name-Tuple{SBMLModel, String}"><code>COBREXA.reaction_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_name(model::SBMLModel, rid::String) -&gt; Union{Nothing, String}
</code></pre><p>Return the name of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L315">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_notes-Tuple{SBMLModel, String}" href="#COBREXA.reaction_notes-Tuple{SBMLModel, String}"><code>COBREXA.reaction_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_notes(model::SBMLModel, rid::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the notes about reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L362">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{SBMLModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{SBMLModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(m::SBMLModel, rid::String) -&gt; Dict{String, Float64}
</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L298">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{SBMLModel}" href="#COBREXA.reactions-Tuple{SBMLModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::SBMLModel) -&gt; Vector{String}
</code></pre><p>Get reactions from a <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L41">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{SBMLModel}" href="#COBREXA.stoichiometry-Tuple{SBMLModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::SBMLModel) -&gt; SparseArrays.SparseMatrixCSC{Float64, Int64}
</code></pre><p>Recreate the stoichiometry matrix from the <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/SBMLModel.jl#L69">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Serialized" href="#COBREXA.Serialized"><code>COBREXA.Serialized</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct Serialized{M} &lt;: ModelWrapper</code></pre><p>A meta-model that represents a model that is serialized on the disk. The internal model will be loaded on-demand by using any accessor, or by calling <a href="#COBREXA.precache!-Tuple{MetabolicModel}"><code>precache!</code></a> directly.</p><p><strong>Fields</strong></p><ul><li><p><code>m::Union{Nothing, M} where M</code></p></li><li><p><code>filename::String</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Serialized.jl#L2">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.precache!-Tuple{Serialized}" href="#COBREXA.precache!-Tuple{Serialized}"><code>COBREXA.precache!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">precache!(model::Serialized)
</code></pre><p>Load the <code>Serialized</code> model from disk in case it&#39;s not alreadly loaded.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Serialized.jl#L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.unwrap_model-Tuple{Serialized}" href="#COBREXA.unwrap_model-Tuple{Serialized}"><code>COBREXA.unwrap_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">unwrap_model(m::Serialized) -&gt; Union{Nothing, M} where M
</code></pre><p>Unwrap the serialized model (precaching it transparently).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/Serialized.jl#L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.StandardModel" href="#COBREXA.StandardModel"><code>COBREXA.StandardModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct StandardModel &lt;: MetabolicModel</code></pre><p><code>StandardModel</code> is used to store a constraint based metabolic model with meta-information.  Meta-information is defined as annotation details, which include gene-reaction-rules, formulas, etc.</p><p>This model type seeks to keep as much meta-information as possible, as opposed to <code>CoreModel</code> and <code>CoreModelCoupled</code>, which keep the bare neccessities only. When merging models and keeping meta-information is important, use this as the model type.  If meta-information is not important, use the more efficient core model types.  See <a href="#COBREXA.CoreModel"><code>CoreModel</code></a> and <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> for comparison.</p><p>In this model, reactions, metabolites, and genes are stored in ordered dictionaries indexed by each struct&#39;s <code>id</code> field.  For example, <code>model.reactions[&quot;rxn1_id&quot;]</code> returns a <code>Reaction</code> where the field <code>id</code> equals <code>&quot;rxn1_id&quot;</code>.  This makes adding and removing reactions efficient.</p><p>Note that the stoichiometric matrix (or any other core data, e.g. flux bounds) is not stored directly as in <code>CoreModel</code>.  When this model type is used in analysis functions, these core data structures are built from scratch each time an analysis function is called.  This can cause performance issues if you run many small analysis functions sequentially.  Consider using the core model types if performance is critical.</p><p>See also: <a href="#COBREXA.Reaction"><code>Reaction</code></a>, <a href="#COBREXA.Metabolite"><code>Metabolite</code></a>, <a href="#COBREXA.Gene"><code>Gene</code></a></p><p><strong>Example</strong></p><pre><code class="language-none">model = load_model(StandardModel, &quot;my_model.json&quot;)
keys(model.reactions)</code></pre><p><strong>Fields</strong></p><ul><li><p><code>id::String</code></p></li><li><p><code>reactions::OrderedCollections.OrderedDict{String, Reaction}</code></p></li><li><p><code>metabolites::OrderedCollections.OrderedDict{String, Metabolite}</code></p></li><li><p><code>genes::OrderedCollections.OrderedDict{String, Gene}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{StandardModel}, MetabolicModel}" href="#Base.convert-Tuple{Type{StandardModel}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">convert(_::Type{StandardModel}, model::MetabolicModel) -&gt; MetabolicModel
</code></pre><p>Convert any <code>MetabolicModel</code> into a <code>StandardModel</code>. Note, some data loss may occur since only the generic interface is used during the conversion process.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L309">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{StandardModel}" href="#COBREXA.balance-Tuple{StandardModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(model::StandardModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p>Return the balance of the linear problem, i.e. b in Sv = 0 where S is the stoichiometric matrix and v is the flux vector.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L167">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{StandardModel}" href="#COBREXA.bounds-Tuple{StandardModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::StandardModel) -&gt; Tuple{Vector{Float64}, Vector{Float64}}
</code></pre><p>Return the lower and upper bounds, respectively, for reactions in <code>model</code>. Order matches that of the reaction ids returned in <code>reactions()</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L158">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_annotations-Tuple{StandardModel, String}" href="#COBREXA.gene_annotations-Tuple{StandardModel, String}"><code>COBREXA.gene_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_annotations(model::StandardModel, id::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the annotation associated with gene <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L254">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_name-Tuple{StandardModel, String}" href="#COBREXA.gene_name-Tuple{StandardModel, String}"><code>COBREXA.gene_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_name(m::StandardModel, gid::String) -&gt; Union{Nothing, String}
</code></pre><p>Return the name of gene with ID <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L302">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_notes-Tuple{StandardModel, String}" href="#COBREXA.gene_notes-Tuple{StandardModel, String}"><code>COBREXA.gene_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_notes(model::StandardModel, id::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the notes associated with gene <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L246">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{StandardModel}" href="#COBREXA.genes-Tuple{StandardModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(model::StandardModel) -&gt; Vector{String}
</code></pre><p>Return a vector of gene id strings in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L86">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.lower_bounds-Tuple{StandardModel}" href="#COBREXA.lower_bounds-Tuple{StandardModel}"><code>COBREXA.lower_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">lower_bounds(model::StandardModel) -&gt; Vector{Float64}
</code></pre><p>Return the lower bounds for all reactions in <code>model</code> in sparse format.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L141">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_annotations-Tuple{StandardModel, String}" href="#COBREXA.metabolite_annotations-Tuple{StandardModel, String}"><code>COBREXA.metabolite_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_annotations(model::StandardModel, id::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the annotation associated with metabolite <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L237">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{StandardModel, String}" href="#COBREXA.metabolite_charge-Tuple{StandardModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::StandardModel, id::String) -&gt; Union{Nothing, Int64}
</code></pre><p>Return the charge associated with metabolite <code>id</code> in <code>model</code>. Return nothing if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L201">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{StandardModel, String}" href="#COBREXA.metabolite_compartment-Tuple{StandardModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(model::StandardModel, id::String) -&gt; Union{Nothing, String}
</code></pre><p>Return compartment associated with metabolite <code>id</code> in <code>model</code>. Return <code>nothing</code> if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L210">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{StandardModel, String}" href="#COBREXA.metabolite_formula-Tuple{StandardModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(model::StandardModel, id::String) -&gt; Union{Nothing, Dict{String, Int64}}
</code></pre><p>Return the formula of reaction <code>id</code> in <code>model</code>. Return <code>nothing</code> if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L192">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_name-Tuple{StandardModel, String}" href="#COBREXA.metabolite_name-Tuple{StandardModel, String}"><code>COBREXA.metabolite_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_name(m::StandardModel, mid::String) -&gt; Union{Nothing, String}
</code></pre><p>Return the name of metabolite with ID <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L295">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_notes-Tuple{StandardModel, String}" href="#COBREXA.metabolite_notes-Tuple{StandardModel, String}"><code>COBREXA.metabolite_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_notes(model::StandardModel, id::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the notes associated with metabolite <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L228">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{StandardModel}" href="#COBREXA.metabolites-Tuple{StandardModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(model::StandardModel) -&gt; Vector{String}
</code></pre><p>Return a vector of metabolite id strings contained in <code>model</code>. The order of metabolite strings returned here matches the order used to construct the stoichiometric matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L70">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_genes-Tuple{StandardModel}" href="#COBREXA.n_genes-Tuple{StandardModel}"><code>COBREXA.n_genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_genes(model::StandardModel) -&gt; Int64
</code></pre><p>Return the number of genes in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L93">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_metabolites-Tuple{StandardModel}" href="#COBREXA.n_metabolites-Tuple{StandardModel}"><code>COBREXA.n_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_metabolites(model::StandardModel) -&gt; Int64
</code></pre><p>Return the number of metabolites in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L79">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_reactions-Tuple{StandardModel}" href="#COBREXA.n_reactions-Tuple{StandardModel}"><code>COBREXA.n_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_reactions(model::StandardModel) -&gt; Int64
</code></pre><p>Return the number of reactions contained in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L62">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{StandardModel}" href="#COBREXA.objective-Tuple{StandardModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::StandardModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p>Return sparse objective vector for <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L175">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_annotations-Tuple{StandardModel, String}" href="#COBREXA.reaction_annotations-Tuple{StandardModel, String}"><code>COBREXA.reaction_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_annotations(model::StandardModel, id::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the annotation associated with reaction <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L271">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{StandardModel, String}" href="#COBREXA.reaction_gene_association-Tuple{StandardModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(model::StandardModel, id::String) -&gt; Union{Nothing, Vector{Vector{String}}}
</code></pre><p>Return the gene reaction rule in string format for reaction with <code>id</code> in <code>model</code>. Return <code>nothing</code> if not available.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L183">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_name-Tuple{StandardModel, String}" href="#COBREXA.reaction_name-Tuple{StandardModel, String}"><code>COBREXA.reaction_name</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_name(m::StandardModel, rid::String) -&gt; Union{Nothing, String}
</code></pre><p>Return the name of reaction with ID <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L288">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_notes-Tuple{StandardModel, String}" href="#COBREXA.reaction_notes-Tuple{StandardModel, String}"><code>COBREXA.reaction_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_notes(model::StandardModel, id::String) -&gt; Dict{String, Vector{String}}
</code></pre><p>Return the notes associated with reaction <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L263">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{StandardModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{StandardModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(m::StandardModel, rid::String) -&gt; Dict{String, Float64}
</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L280">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_subsystem-Tuple{StandardModel, String}" href="#COBREXA.reaction_subsystem-Tuple{StandardModel, String}"><code>COBREXA.reaction_subsystem</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_subsystem(model::StandardModel, id::String) -&gt; Union{Nothing, String}
</code></pre><p>Return the subsystem associated with reaction <code>id</code> in <code>model</code>. Return <code>nothing</code> if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L219">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{StandardModel}" href="#COBREXA.reactions-Tuple{StandardModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::StandardModel) -&gt; Vector{String}
</code></pre><p>Return a vector of reaction id strings contained in <code>model</code>. The order of reaction ids returned here matches the order used to construct the stoichiometric matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{StandardModel}" href="#COBREXA.stoichiometry-Tuple{StandardModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::StandardModel) -&gt; SparseArrays.SparseMatrixCSC{Float64, Int64}
</code></pre><p>Return the stoichiometric matrix associated with <code>model</code> in sparse format.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L100">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.upper_bounds-Tuple{StandardModel}" href="#COBREXA.upper_bounds-Tuple{StandardModel}"><code>COBREXA.upper_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">upper_bounds(model::StandardModel) -&gt; Vector{Float64}
</code></pre><p>Return the upper bounds for all reactions in <code>model</code> in sparse format. Order matches that of the reaction ids returned in <code>reactions()</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/StandardModel.jl#L149">source</a></section></article><h2 id="Model-type-wrappers"><a class="docs-heading-anchor" href="#Model-type-wrappers">Model type wrappers</a><a id="Model-type-wrappers-1"></a><a class="docs-heading-anchor-permalink" href="#Model-type-wrappers" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.ExpressionLimitedModel" href="#COBREXA.ExpressionLimitedModel"><code>COBREXA.ExpressionLimitedModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct ExpressionLimitedModel &lt;: ModelWrapper</code></pre><p><a href="#COBREXA.ExpressionLimitedModel"><code>ExpressionLimitedModel</code></a> follows the methodology of the E-Flux algorithm to constraint the flux through the reactions in order to simulate the limited expression of genes (and thus the limited availability of the gene products).</p><p>Use <a href="../analysis/#COBREXA.make_expression_limited_model-Tuple{MetabolicModel}"><code>make_expression_limited_model</code></a> or <a href="../reconstruction/#COBREXA.with_expression_limits-Tuple{}"><code>with_expression_limits</code></a> to construct the models.</p><p>E-Flux algorithm is closer described by: <em>Colijn, Caroline, Aaron Brandes, Jeremy Zucker, Desmond S. Lun, Brian Weiner, Maha R. Farhat, Tan-Yun Cheng, D. Branch Moody, Megan Murray, and James E. Galagan. &quot;Interpreting expression data with metabolic flux models: predicting Mycobacterium tuberculosis mycolic acid production.&quot; PLoS computational biology 5, no. 8 (2009): e1000489</em>.</p><p><strong>Fields</strong></p><ul><li><p><code>relative_expression::Dict{String, Float64}</code></p><p>Relative gene expression w.r.t. to some chosen reference; the normalization and scale of the values should match the expectations of the <code>bounding_function</code>.</p></li></ul><ul><li><p><code>inner::MetabolicModel</code></p><p>The wrapped model.</p></li><li><p><code>bounding_function::Function</code></p><p>Function used to calculate the new reaction bounds from expression. In <a href="../analysis/#COBREXA.make_expression_limited_model-Tuple{MetabolicModel}"><code>make_expression_limited_model</code></a> this defaults to <a href="../utils/#COBREXA.expression_probabilistic_bounds-Tuple{Dict{String, Float64}, Union{Nothing, Vector{Vector{String}}}}"><code>expression_probabilistic_bounds</code></a>.</p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/ExpressionLimitedModel.jl#L2-L20">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.GeckoModel" href="#COBREXA.GeckoModel"><code>COBREXA.GeckoModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct GeckoModel &lt;: ModelWrapper</code></pre><p>A model with complex enzyme concentration and capacity bounds, as described in <em>Sánchez, Benjamín J., et al. &quot;Improving the phenotype predictions of a yeast genome-scale metabolic model by incorporating enzymatic constraints.&quot; Molecular systems biology 13.8 (2017): 935.</em></p><p>Use <a href="../analysis/#COBREXA.make_gecko_model-Tuple{MetabolicModel}"><code>make_gecko_model</code></a> or <a href="../reconstruction/#COBREXA.with_gecko-Tuple{}"><code>with_gecko</code></a> to construct this kind of model.</p><p>The model wraps another &quot;internal&quot; model, and adds following modifications:</p><ul><li>enzymatic reactions with known enzyme information are split into multiple forward and reverse variants for each isozyme,</li><li>reaction coupling is added to ensure the groups of isozyme reactions obey the global reaction flux bounds from the original model,</li><li>gene concentrations specified by each reaction and its gene product stoichiometry, can constrained by the user to reflect measurements, such as from mass spectrometry,</li><li>additional coupling is added to simulate total masses of different proteins grouped by type (e.g., membrane-bound and free-floating proteins), which can be again constrained by the user (this is slightly generalized from original GECKO algorithm, which only considers a single group of indiscernible proteins).</li></ul><p>The structure contains fields <code>columns</code> that describe the contents of the stoichiometry matrix columns, <code>coupling_row_reaction</code>, <code>coupling_row_gene_product</code> and <code>coupling_row_mass_group</code> that describe correspondence of the coupling rows to original model and determine the coupling bounds (note: the coupling for gene product is actually added to stoichiometry, not in <a href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>coupling</code></a>), and <code>inner</code>, which is the original wrapped model. The <code>objective</code> of the model includes also the extra columns for individual genes, as held by <code>coupling_row_gene_product</code>.</p><p>Implementation exposes the split reactions (available as <code>reactions(model)</code>), but retains the original &quot;simple&quot; reactions accessible by <a href="#COBREXA.fluxes-Tuple{MetabolicModel}"><code>fluxes</code></a>. The related constraints are implemented using <a href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>coupling</code></a> and <a href="#COBREXA.coupling_bounds-Tuple{CoreCoupling}"><code>coupling_bounds</code></a>.</p><p><strong>Fields</strong></p><ul><li><p><code>objective::SparseArrays.SparseVector{Float64, Int64}</code></p></li><li><p><code>columns::Vector{COBREXA._gecko_reaction_column}</code></p></li><li><p><code>coupling_row_reaction::Vector{Int64}</code></p></li><li><p><code>coupling_row_gene_product::Vector{Tuple{Int64, Tuple{Float64, Float64}}}</code></p></li><li><p><code>coupling_row_mass_group::Vector{COBREXA._gecko_capacity}</code></p></li><li><p><code>inner::MetabolicModel</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L35">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._gecko_capacity" href="#COBREXA._gecko_capacity"><code>COBREXA._gecko_capacity</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct _gecko_capacity</code></pre><p>A helper struct that contains the gene product capacity terms organized by the grouping type, e.g. metabolic or membrane groups etc.</p><p><strong>Fields</strong></p><ul><li><p><code>group_id::String</code></p></li><li><p><code>gene_product_idxs::Vector{Int64}</code></p></li><li><p><code>gene_product_molar_masses::Vector{Float64}</code></p></li><li><p><code>group_upper_bound::Float64</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L19">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._gecko_reaction_column" href="#COBREXA._gecko_reaction_column"><code>COBREXA._gecko_reaction_column</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct _gecko_reaction_column</code></pre><p>A helper type for describing the contents of <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a>s.</p><p><strong>Fields</strong></p><ul><li><p><code>reaction_idx::Int64</code></p></li><li><p><code>isozyme_idx::Int64</code></p></li><li><p><code>direction::Int64</code></p></li><li><p><code>reaction_coupling_row::Int64</code></p></li><li><p><code>lb::Float64</code></p></li><li><p><code>ub::Float64</code></p></li><li><p><code>gene_product_coupling::Vector{Tuple{Int64, Float64}}</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L1">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{GeckoModel}" href="#COBREXA.balance-Tuple{GeckoModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(model::GeckoModel) -&gt; Any
</code></pre><p>Return the balance of the reactions in the inner model, concatenated with a vector of zeros representing the enzyme balance of a <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L225">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{GeckoModel}" href="#COBREXA.bounds-Tuple{GeckoModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::GeckoModel) -&gt; Tuple{Vector{Float64}, Vector{Float64}}
</code></pre><p>Return variable bounds for <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L142">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{GeckoModel}" href="#COBREXA.coupling-Tuple{GeckoModel}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(model::GeckoModel) -&gt; Any
</code></pre><p>Return the coupling of <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a>. That combines the coupling of the wrapped model, coupling for split (arm) reactions, and the coupling for the total enzyme capacity.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L173">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{GeckoModel}" href="#COBREXA.coupling_bounds-Tuple{GeckoModel}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(model::GeckoModel) -&gt; Tuple{Any, Any}
</code></pre><p>The coupling bounds for <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a> (refer to <a href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>coupling</code></a> for details).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L202">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{GeckoModel}" href="#COBREXA.genes-Tuple{GeckoModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(model::GeckoModel) -&gt; Any
</code></pre><p>Return the gene ids of genes that have enzymatic constraints associated with them.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L241">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{GeckoModel}" href="#COBREXA.metabolites-Tuple{GeckoModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(model::GeckoModel) -&gt; Any
</code></pre><p>Return the ids of all metabolites, both real and pseudo, for a <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L249">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_coupling_constraints-Tuple{GeckoModel}" href="#COBREXA.n_coupling_constraints-Tuple{GeckoModel}"><code>COBREXA.n_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_coupling_constraints(model::GeckoModel) -&gt; Any
</code></pre><p>Count the coupling constraints in <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a> (refer to <a href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>coupling</code></a> for details).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L191">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_genes-Tuple{GeckoModel}" href="#COBREXA.n_genes-Tuple{GeckoModel}"><code>COBREXA.n_genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_genes(model::GeckoModel) -&gt; Int64
</code></pre><p>Return the number of genes that have enzymatic constraints associated with them.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L234">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_metabolites-Tuple{GeckoModel}" href="#COBREXA.n_metabolites-Tuple{GeckoModel}"><code>COBREXA.n_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_metabolites(model::GeckoModel) -&gt; Any
</code></pre><p>Return the number of metabolites, both real and pseudo, for a <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L256">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_reactions-Tuple{GeckoModel}" href="#COBREXA.n_reactions-Tuple{GeckoModel}"><code>COBREXA.n_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_reactions(model::GeckoModel) -&gt; Int64
</code></pre><p>Returns the number of all irreversible reactions in <code>model</code> as well as the number of gene products that take part in enzymatic reactions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L134">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{GeckoModel}" href="#COBREXA.objective-Tuple{GeckoModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::GeckoModel) -&gt; SparseArrays.SparseVector{Float64, Int64}
</code></pre><p>Return the objective of the <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a>. Note, the objective is with respect to the internal variables, i.e. <a href="#COBREXA.reactions-Tuple{CoreModel}"><code>reactions(model)</code></a>, which are the unidirectional reactions and the genes involved in enzymatic reactions that have kinetic data.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L105">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_flux-Tuple{GeckoModel}" href="#COBREXA.reaction_flux-Tuple{GeckoModel}"><code>COBREXA.reaction_flux</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_flux(model::GeckoModel) -&gt; Any
</code></pre><p>Get the mapping of the reaction rates in <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a> to the original fluxes in the wrapped model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L159">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{GeckoModel}" href="#COBREXA.reactions-Tuple{GeckoModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::GeckoModel) -&gt; Any
</code></pre><p>Returns the internal reactions in a <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a> (these may be split to forward- and reverse-only parts with different isozyme indexes; reactions IDs are mangled accordingly with suffixes).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L115">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{GeckoModel}" href="#COBREXA.stoichiometry-Tuple{GeckoModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::GeckoModel) -&gt; Any
</code></pre><p>Return a stoichiometry of the <a href="#COBREXA.GeckoModel"><code>GeckoModel</code></a>. The enzymatic reactions are split into unidirectional forward and reverse ones, each of which may have multiple variants per isozyme.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/GeckoModel.jl#L89">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.SMomentModel" href="#COBREXA.SMomentModel"><code>COBREXA.SMomentModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct SMomentModel &lt;: ModelWrapper</code></pre><p>An enzyme-capacity-constrained model using sMOMENT algorithm, as described by <em>Bekiaris, Pavlos Stephanos, and Steffen Klamt, &quot;Automatic construction of metabolic models with enzyme constraints&quot; BMC bioinformatics, 2020</em>.</p><p>Use <a href="../analysis/#COBREXA.make_smoment_model-Tuple{MetabolicModel}"><code>make_smoment_model</code></a> or <a href="../reconstruction/#COBREXA.with_smoment-Tuple{}"><code>with_smoment</code></a> to construct the models.</p><p>The model is constructed as follows:</p><ul><li>stoichiometry of the original model is retained as much as possible, but enzymatic reations are split into forward and reverse parts (marked by a suffix like <code>...#forward</code> and <code>...#reverse</code>),</li><li>coupling is added to simulate a virtual metabolite &quot;enzyme capacity&quot;, which is consumed by all enzymatic reactions at a rate given by enzyme mass divided by the corresponding kcat,</li><li>the total consumption of the enzyme capacity is constrained to a fixed maximum.</li></ul><p>The <code>SMomentModel</code> structure contains a worked-out representation of the optimization problem atop a wrapped <a href="#COBREXA.MetabolicModel"><code>MetabolicModel</code></a>, in particular the separation of certain reactions into unidirectional forward and reverse parts, an &quot;enzyme capacity&quot; required for each reaction, and the value of the maximum capacity constraint. Original coupling in the inner model is retained.</p><p>In the structure, the field <code>columns</code> describes the correspondence of stoichiometry columns to the stoichiometry and data of the internal wrapped model, and <code>total_enzyme_capacity</code> is the total bound on the enzyme capacity consumption as specified in sMOMENT algorithm.</p><p>This implementation allows easy access to fluxes from the split reactions (available in <code>reactions(model)</code>), while the original &quot;simple&quot; reactions from the wrapped model are retained as <a href="#COBREXA.fluxes-Tuple{MetabolicModel}"><code>fluxes</code></a>. All additional constraints are implemented using <a href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>coupling</code></a> and <a href="#COBREXA.coupling_bounds-Tuple{CoreCoupling}"><code>coupling_bounds</code></a>.</p><p><strong>Fields</strong></p><ul><li><p><code>columns::Vector{COBREXA._smoment_column}</code></p></li><li><p><code>total_enzyme_capacity::Float64</code></p></li><li><p><code>inner::MetabolicModel</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L18">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._smoment_column" href="#COBREXA._smoment_column"><code>COBREXA._smoment_column</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct _smoment_column</code></pre><p>A helper type that describes the contents of <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a>s.</p><p><strong>Fields</strong></p><ul><li><p><code>reaction_idx::Int64</code></p></li><li><p><code>direction::Int64</code></p></li><li><p><code>lb::Float64</code></p></li><li><p><code>ub::Float64</code></p></li><li><p><code>capacity_required::Float64</code></p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L2">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{SMomentModel}" href="#COBREXA.bounds-Tuple{SMomentModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::SMomentModel) -&gt; Tuple{Vector{Float64}, Vector{Float64}}
</code></pre><p>Return the variable bounds for <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L104">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{SMomentModel}" href="#COBREXA.coupling-Tuple{SMomentModel}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(model::SMomentModel) -&gt; Any
</code></pre><p>Return the coupling of <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a>. That combines the coupling of the wrapped model, coupling for split reactions, and the coupling for the total enzyme capacity.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L121">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{SMomentModel}" href="#COBREXA.coupling_bounds-Tuple{SMomentModel}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(model::SMomentModel) -&gt; Tuple{Any, Any}
</code></pre><p>The coupling bounds for <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a> (refer to <a href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>coupling</code></a> for details).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L141">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_coupling_constraints-Tuple{SMomentModel}" href="#COBREXA.n_coupling_constraints-Tuple{SMomentModel}"><code>COBREXA.n_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_coupling_constraints(model::SMomentModel) -&gt; Any
</code></pre><p>Count the coupling constraints in <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a> (refer to <a href="#COBREXA.coupling-Tuple{CoreCoupling}"><code>coupling</code></a> for details).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L133">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_reactions-Tuple{SMomentModel}" href="#COBREXA.n_reactions-Tuple{SMomentModel}"><code>COBREXA.n_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_reactions(model::SMomentModel) -&gt; Int64
</code></pre><p>The number of reactions (including split ones) in <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L97">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{SMomentModel}" href="#COBREXA.objective-Tuple{SMomentModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::SMomentModel) -&gt; Any
</code></pre><p>Reconstruct an objective of the <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L75">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_flux-Tuple{SMomentModel}" href="#COBREXA.reaction_flux-Tuple{SMomentModel}"><code>COBREXA.reaction_flux</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_flux(model::SMomentModel) -&gt; Any
</code></pre><p>Get the mapping of the reaction rates in <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a> to the original fluxes in the wrapped model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L112">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{SMomentModel}" href="#COBREXA.reactions-Tuple{SMomentModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::SMomentModel) -&gt; Vector
</code></pre><p>Returns the internal reactions in a <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a> (these may be split to forward- and reverse-only parts; reactions IDs are mangled accordingly with suffixes).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L82">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{SMomentModel}" href="#COBREXA.stoichiometry-Tuple{SMomentModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::SMomentModel) -&gt; Any
</code></pre><p>Return a stoichiometry of the <a href="#COBREXA.SMomentModel"><code>SMomentModel</code></a>. The enzymatic reactions are split into unidirectional forward and reverse ones.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/lcsb-biocore/COBREXA.jl/blob/99b30024a2eeb42f0c3ca14fa4d5523a28ae769a/src/base/types/wrappers/SMomentModel.jl#L66">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../reconstruction/">« Model construction functions</a><a class="docs-footer-nextpage" href="../utils/">Utilities »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Thursday 11 May 2023 10:45">Thursday 11 May 2023</span>. Using Julia version 1.8.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
